

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>API Reference &mdash; pyisx 2.0.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=b86133f3" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/theme.css?v=547657ed" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=51b770b3"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Examples" href="examples.html" />
    <link rel="prev" title="Installation" href="installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            pyisx
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="overview.html">Overview</a></li>
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">API Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#isx.CellSet"><code class="docutils literal notranslate"><span class="pre">CellSet</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.flush"><code class="docutils literal notranslate"><span class="pre">CellSet.flush()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.get_acquisition_info"><code class="docutils literal notranslate"><span class="pre">CellSet.get_acquisition_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.get_cell_image_data"><code class="docutils literal notranslate"><span class="pre">CellSet.get_cell_image_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.get_cell_name"><code class="docutils literal notranslate"><span class="pre">CellSet.get_cell_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.get_cell_status"><code class="docutils literal notranslate"><span class="pre">CellSet.get_cell_status()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.get_cell_trace_data"><code class="docutils literal notranslate"><span class="pre">CellSet.get_cell_trace_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.read"><code class="docutils literal notranslate"><span class="pre">CellSet.read()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.set_cell_data"><code class="docutils literal notranslate"><span class="pre">CellSet.set_cell_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.set_cell_status"><code class="docutils literal notranslate"><span class="pre">CellSet.set_cell_status()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.CellSet.write"><code class="docutils literal notranslate"><span class="pre">CellSet.write()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.Duration"><code class="docutils literal notranslate"><span class="pre">Duration</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.Duration.from_msecs"><code class="docutils literal notranslate"><span class="pre">Duration.from_msecs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Duration.from_secs"><code class="docutils literal notranslate"><span class="pre">Duration.from_secs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Duration.from_usecs"><code class="docutils literal notranslate"><span class="pre">Duration.from_usecs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Duration.to_msecs"><code class="docutils literal notranslate"><span class="pre">Duration.to_msecs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Duration.to_secs"><code class="docutils literal notranslate"><span class="pre">Duration.to_secs()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Duration.to_usecs"><code class="docutils literal notranslate"><span class="pre">Duration.to_usecs()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.Enum"><code class="docutils literal notranslate"><span class="pre">Enum</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.EventSet"><code class="docutils literal notranslate"><span class="pre">EventSet</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.EventSet.flush"><code class="docutils literal notranslate"><span class="pre">EventSet.flush()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.EventSet.get_acquisition_info"><code class="docutils literal notranslate"><span class="pre">EventSet.get_acquisition_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.EventSet.get_cell_data"><code class="docutils literal notranslate"><span class="pre">EventSet.get_cell_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.EventSet.get_cell_index"><code class="docutils literal notranslate"><span class="pre">EventSet.get_cell_index()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.EventSet.get_cell_name"><code class="docutils literal notranslate"><span class="pre">EventSet.get_cell_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.EventSet.read"><code class="docutils literal notranslate"><span class="pre">EventSet.read()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.EventSet.set_cell_data"><code class="docutils literal notranslate"><span class="pre">EventSet.set_cell_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.EventSet.write"><code class="docutils literal notranslate"><span class="pre">EventSet.write()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.Fraction"><code class="docutils literal notranslate"><span class="pre">Fraction</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.Fraction.as_integer_ratio"><code class="docutils literal notranslate"><span class="pre">Fraction.as_integer_ratio()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Fraction.from_decimal"><code class="docutils literal notranslate"><span class="pre">Fraction.from_decimal()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Fraction.from_float"><code class="docutils literal notranslate"><span class="pre">Fraction.from_float()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Fraction.limit_denominator"><code class="docutils literal notranslate"><span class="pre">Fraction.limit_denominator()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.GpioSet"><code class="docutils literal notranslate"><span class="pre">GpioSet</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.GpioSet.get_acquisition_info"><code class="docutils literal notranslate"><span class="pre">GpioSet.get_acquisition_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.GpioSet.get_channel_data"><code class="docutils literal notranslate"><span class="pre">GpioSet.get_channel_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.GpioSet.get_channel_index"><code class="docutils literal notranslate"><span class="pre">GpioSet.get_channel_index()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.GpioSet.get_channel_name"><code class="docutils literal notranslate"><span class="pre">GpioSet.get_channel_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.GpioSet.read"><code class="docutils literal notranslate"><span class="pre">GpioSet.read()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.Image"><code class="docutils literal notranslate"><span class="pre">Image</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.Image.get_data"><code class="docutils literal notranslate"><span class="pre">Image.get_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Image.read"><code class="docutils literal notranslate"><span class="pre">Image.read()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Image.write"><code class="docutils literal notranslate"><span class="pre">Image.write()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.Movie"><code class="docutils literal notranslate"><span class="pre">Movie</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.Movie.flush"><code class="docutils literal notranslate"><span class="pre">Movie.flush()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Movie.get_acquisition_info"><code class="docutils literal notranslate"><span class="pre">Movie.get_acquisition_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Movie.get_frame_data"><code class="docutils literal notranslate"><span class="pre">Movie.get_frame_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Movie.get_frame_timestamp"><code class="docutils literal notranslate"><span class="pre">Movie.get_frame_timestamp()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Movie.read"><code class="docutils literal notranslate"><span class="pre">Movie.read()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Movie.set_frame_data"><code class="docutils literal notranslate"><span class="pre">Movie.set_frame_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Movie.write"><code class="docutils literal notranslate"><span class="pre">Movie.write()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.Spacing"><code class="docutils literal notranslate"><span class="pre">Spacing</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.Time"><code class="docutils literal notranslate"><span class="pre">Time</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.Time.to_datetime"><code class="docutils literal notranslate"><span class="pre">Time.to_datetime()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.Timing"><code class="docutils literal notranslate"><span class="pre">Timing</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.Timing.get_offsets_since_start"><code class="docutils literal notranslate"><span class="pre">Timing.get_offsets_since_start()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Timing.get_valid_samples"><code class="docutils literal notranslate"><span class="pre">Timing.get_valid_samples()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.Timing.get_valid_samples_mask"><code class="docutils literal notranslate"><span class="pre">Timing.get_valid_samples_mask()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.VesselSet"><code class="docutils literal notranslate"><span class="pre">VesselSet</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.VesselSetType"><code class="docutils literal notranslate"><span class="pre">VesselSet.VesselSetType</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.flush"><code class="docutils literal notranslate"><span class="pre">VesselSet.flush()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_acquisition_info"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_acquisition_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_center_trace_data"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_center_trace_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_correlations_data"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_correlations_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_direction_trace_data"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_direction_trace_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_image_data"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_image_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_line_data"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_line_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_name"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_set_type"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_set_type()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_status"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_status()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.get_vessel_trace_data"><code class="docutils literal notranslate"><span class="pre">VesselSet.get_vessel_trace_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.has_correlation_heatmaps"><code class="docutils literal notranslate"><span class="pre">VesselSet.has_correlation_heatmaps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.read"><code class="docutils literal notranslate"><span class="pre">VesselSet.read()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.set_rbc_velocity_data"><code class="docutils literal notranslate"><span class="pre">VesselSet.set_rbc_velocity_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.set_vessel_diameter_data"><code class="docutils literal notranslate"><span class="pre">VesselSet.set_vessel_diameter_data()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.set_vessel_status"><code class="docutils literal notranslate"><span class="pre">VesselSet.set_vessel_status()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#isx.VesselSet.write"><code class="docutils literal notranslate"><span class="pre">VesselSet.write()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#isx.align_start_times"><code class="docutils literal notranslate"><span class="pre">align_start_times()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.convert_type_numpy_array"><code class="docutils literal notranslate"><span class="pre">convert_type_numpy_array()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.create_cell_map"><code class="docutils literal notranslate"><span class="pre">create_cell_map()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_aligned_timestamps"><code class="docutils literal notranslate"><span class="pre">export_aligned_timestamps()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_cell_set_to_csv_tiff"><code class="docutils literal notranslate"><span class="pre">export_cell_set_to_csv_tiff()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_ethovision_data_with_isxb_timestamps"><code class="docutils literal notranslate"><span class="pre">export_ethovision_data_with_isxb_timestamps()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_event_set_to_csv"><code class="docutils literal notranslate"><span class="pre">export_event_set_to_csv()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_gpio_set_to_csv"><code class="docutils literal notranslate"><span class="pre">export_gpio_set_to_csv()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_gpio_to_isxd"><code class="docutils literal notranslate"><span class="pre">export_gpio_to_isxd()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_image_to_isxd_tiff"><code class="docutils literal notranslate"><span class="pre">export_image_to_isxd_tiff()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_image_to_tiff"><code class="docutils literal notranslate"><span class="pre">export_image_to_tiff()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_isxd_image_to_tiff"><code class="docutils literal notranslate"><span class="pre">export_isxd_image_to_tiff()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_movie_timestamps_to_csv"><code class="docutils literal notranslate"><span class="pre">export_movie_timestamps_to_csv()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_movie_to_mp4"><code class="docutils literal notranslate"><span class="pre">export_movie_to_mp4()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_movie_to_nwb"><code class="docutils literal notranslate"><span class="pre">export_movie_to_nwb()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_movie_to_tiff"><code class="docutils literal notranslate"><span class="pre">export_movie_to_tiff()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_nvision_movie_tracking_frame_data_to_csv"><code class="docutils literal notranslate"><span class="pre">export_nvision_movie_tracking_frame_data_to_csv()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_nvision_movie_tracking_zone_data_to_csv"><code class="docutils literal notranslate"><span class="pre">export_nvision_movie_tracking_zone_data_to_csv()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_tiff_to_isxd_image"><code class="docutils literal notranslate"><span class="pre">export_tiff_to_isxd_image()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.export_vessel_set_to_csv_tiff"><code class="docutils literal notranslate"><span class="pre">export_vessel_set_to_csv_tiff()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.get_file_extension"><code class="docutils literal notranslate"><span class="pre">get_file_extension()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.get_file_stem"><code class="docutils literal notranslate"><span class="pre">get_file_stem()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.make_output_file_path"><code class="docutils literal notranslate"><span class="pre">make_output_file_path()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.make_output_file_paths"><code class="docutils literal notranslate"><span class="pre">make_output_file_paths()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.overlay_cell_map_on_image"><code class="docutils literal notranslate"><span class="pre">overlay_cell_map_on_image()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.overlay_cellmaps"><code class="docutils literal notranslate"><span class="pre">overlay_cellmaps()</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#isx.verify_deinterleave"><code class="docutils literal notranslate"><span class="pre">verify_deinterleave()</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="examples.html">Examples</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">FAQ</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">pyisx</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">API Reference</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/reference.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="api-reference">
<span id="reference"></span><h1>API Reference<a class="headerlink" href="#api-reference" title="Link to this heading"></a></h1>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>All hidden modules, classes, functions and members (i.e. those starting with _)
are hidden because they are not part of the public API and their behavior
may change in future versions.
We recommend you do not use them and will not provide support for them.
This includes the <code class="docutils literal notranslate"><span class="pre">_internal</span></code> module.</p>
</div>
<p id="module-isx">The Inscopix Python API package.</p>
<dl class="py class">
<dt class="sig sig-object py" id="isx.CellSet">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">CellSet</span></span><a class="headerlink" href="#isx.CellSet" title="Link to this definition"></a></dt>
<dd><p>A cell set contains the image and trace data associated with components in
a movie, such as cells or regions of interest.</p>
<p>It is always backed by a file in the native <cite>.isxd</cite> format.</p>
<section id="examples">
<h2>Examples<a class="headerlink" href="#examples" title="Link to this heading"></a></h2>
<p>Read an existing cell set from a file and get the image and trace data of
the first cell.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cell_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">CellSet</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP-BP-MC-DFF-PCA-ICA.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">image_0</span> <span class="o">=</span> <span class="n">cell_set</span><span class="o">.</span><span class="n">get_cell_image_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">trace_0</span> <span class="o">=</span> <span class="n">cell_set</span><span class="o">.</span><span class="n">get_cell_trace_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Write a new cell set to a file with the same timing and spacing as an
existing movie, with 3 random cell images and traces.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">movie</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Movie</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cell_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">CellSet</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;cell_set.isxd&#39;</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">timing</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">spacing</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">image</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">cell_set</span><span class="o">.</span><span class="n">spacing</span><span class="o">.</span><span class="n">num_pixels</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">trace</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">cell_set</span><span class="o">.</span><span class="n">timing</span><span class="o">.</span><span class="n">num_samples</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">cell_set</span><span class="o">.</span><span class="n">set_cell_data</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">image</span><span class="p">,</span> <span class="n">trace</span><span class="p">,</span> <span class="s1">&#39;C</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cell_set</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="attributes">
<h2>Attributes<a class="headerlink" href="#attributes" title="Link to this heading"></a></h2>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file that stores this.</p>
</dd>
<dt>mode<span class="classifier">{‘r’, ‘w’}</span></dt><dd><p>The mode the file was opened with.</p>
</dd>
<dt>timing<span class="classifier"><a class="reference internal" href="#isx.Timing" title="isx.Timing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Timing</span></code></a></span></dt><dd><p>The timing of the samples in each cell trace.</p>
</dd>
<dt>spacing<span class="classifier"><a class="reference internal" href="#isx.Spacing" title="isx.Spacing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Spacing</span></code></a></span></dt><dd><p>The spacing of the pixels in each cell image.</p>
</dd>
<dt>num_cells<span class="classifier">int</span></dt><dd><p>The number of cells or components.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.flush" title="Link to this definition"></a></dt>
<dd><p>Flush all meta-data and cell data to the file.</p>
<p>This should be called after setting all cell data of a cell set opened with <a class="reference internal" href="#isx.CellSet.write" title="isx.CellSet.write"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.CellSet.write()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.get_acquisition_info">
<span class="sig-name descname"><span class="pre">get_acquisition_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.get_acquisition_info" title="Link to this definition"></a></dt>
<dd><p>Get information about acquisition that may be stored in some files,
such as nVista 3 movies and data derived from those.</p>
<section id="returns">
<h3>Returns<a class="headerlink" href="#returns" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>dict</dt><dd><p>A dictionary likely parsed from JSON that maps from string keys to variant values.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.get_cell_image_data">
<span class="sig-name descname"><span class="pre">get_cell_image_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.get_cell_image_data" title="Link to this definition"></a></dt>
<dd><p>Get the image data associated with a cell.</p>
<section id="arguments">
<h3>Arguments<a class="headerlink" href="#arguments" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
</dl>
</section>
<section id="id1">
<h3>Returns<a class="headerlink" href="#id1" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The image data in a 2D array.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.get_cell_name">
<span class="sig-name descname"><span class="pre">get_cell_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.get_cell_name" title="Link to this definition"></a></dt>
<dd><section id="id2">
<h3>Arguments<a class="headerlink" href="#id2" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
</dl>
</section>
<section id="id3">
<h3>Returns<a class="headerlink" href="#id3" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>str</dt><dd><p>The name of the indexed cell.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.get_cell_status">
<span class="sig-name descname"><span class="pre">get_cell_status</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.get_cell_status" title="Link to this definition"></a></dt>
<dd><section id="id4">
<h3>Arguments<a class="headerlink" href="#id4" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
</dl>
</section>
<section id="id5">
<h3>Returns<a class="headerlink" href="#id5" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>{‘accepted’, ‘undecided’, ‘rejected’}</dt><dd><p>The status of the indexed cell as a string.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.get_cell_trace_data">
<span class="sig-name descname"><span class="pre">get_cell_trace_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.get_cell_trace_data" title="Link to this definition"></a></dt>
<dd><p>Get the trace data associated with a cell.</p>
<section id="id6">
<h3>Arguments<a class="headerlink" href="#id6" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
</dl>
</section>
<section id="id7">
<h3>Returns<a class="headerlink" href="#id7" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The trace data in a 1D array.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.read">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.read" title="Link to this definition"></a></dt>
<dd><p>Open an existing cell set from a file for reading.</p>
<p>This is a light weight operation that simply reads the meta-data from the cell set,
and does not read any image or trace data.</p>
<section id="id8">
<h3>Arguments<a class="headerlink" href="#id8" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to read.</p>
</dd>
<dt>read_only<span class="classifier">bool</span></dt><dd><p>If true, only allow meta-data and data to be read, otherwise allow some meta-data
to be written (e.g. cell status).</p>
</dd>
</dl>
</section>
<section id="id9">
<h3>Returns<a class="headerlink" href="#id9" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.CellSet" title="isx.CellSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.CellSet</span></code></a></dt><dd><p>The cell set that was read. Meta-data is immediately available.
Image and trace data must be read using <a class="reference internal" href="#isx.CellSet.get_cell_image_data" title="isx.CellSet.get_cell_image_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.CellSet.get_cell_image_data()</span></code></a>
and <a class="reference internal" href="#isx.CellSet.get_cell_trace_data" title="isx.CellSet.get_cell_trace_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.CellSet.get_cell_trace_data()</span></code></a> respectively.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.set_cell_data">
<span class="sig-name descname"><span class="pre">set_cell_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.set_cell_data" title="Link to this definition"></a></dt>
<dd><p>Set the image and trace data of a cell.</p>
<p>Cells must be set in increasing order, otherwise this will error.</p>
<section id="id10">
<h3>Arguments<a class="headerlink" href="#id10" title="Link to this heading"></a></h3>
<dl>
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
<dt>image<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The image data in a 2D array.</p>
</dd>
<dt>trace<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The trace data in a 1D array.</p>
</dd>
<dt>name<span class="classifier">str</span></dt><dd><p>The name of the cell.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.set_cell_status">
<span class="sig-name descname"><span class="pre">set_cell_status</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">status</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.set_cell_status" title="Link to this definition"></a></dt>
<dd><p>Set the status of cell. This will also flush the file.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>As this flushes the file, only use this after all cells have been
written using <a class="reference internal" href="#isx.CellSet.set_cell_data" title="isx.CellSet.set_cell_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.CellSet.set_cell_data()</span></code></a>.</p>
</div>
<section id="id11">
<h3>Arguments<a class="headerlink" href="#id11" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
<dt>status<span class="classifier">{‘accepted’, ‘undecided’, ‘rejected’}</span></dt><dd><p>The desired status of the indexed cell.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.CellSet.write">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timing</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spacing</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.CellSet.write" title="Link to this definition"></a></dt>
<dd><p>Open a new cell set to a file for writing.</p>
<p>This is a light weight operation. It does not write any image or trace data immediately.</p>
<section id="id12">
<h3>Arguments<a class="headerlink" href="#id12" title="Link to this heading"></a></h3>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to write. If it already exists, this will error.</p>
</dd>
<dt>timing<span class="classifier"><a class="reference internal" href="#isx.Timing" title="isx.Timing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Timing</span></code></a></span></dt><dd><p>The timing of the cell set to write. Typically this comes from the movie this
is derived from.</p>
</dd>
<dt>spacing<span class="classifier"><a class="reference internal" href="#isx.Spacing" title="isx.Spacing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Spacing</span></code></a></span></dt><dd><p>The spacing of the movie to write. Typically this comes from the movie this is
derived from.</p>
</dd>
</dl>
</section>
<section id="id13">
<h3>Returns<a class="headerlink" href="#id13" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.CellSet" title="isx.CellSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.CellSet</span></code></a></dt><dd><p>The empty cell set that was written.
Image and trace data must be written with <a class="reference internal" href="#isx.CellSet.set_cell_data" title="isx.CellSet.set_cell_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.CellSet.set_cell_data()</span></code></a>.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.Duration">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">Duration</span></span><a class="headerlink" href="#isx.Duration" title="Link to this definition"></a></dt>
<dd><p>A duration of time.</p>
<section id="id14">
<h2>Examples<a class="headerlink" href="#id14" title="Link to this heading"></a></h2>
<p>Make a period of 50 milliseconds</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">period</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Duration</span><span class="o">.</span><span class="n">from_msecs</span><span class="p">(</span><span class="mi">50</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id15">
<h2>Attributes<a class="headerlink" href="#id15" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>secs_float<span class="classifier">float</span></dt><dd><p>The duration in seconds as a floating point number.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="isx.Duration.from_msecs">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_msecs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">msecs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Duration.from_msecs" title="Link to this definition"></a></dt>
<dd><p>Make a duration from a number of milliseconds.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Duration.from_secs">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_secs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">secs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Duration.from_secs" title="Link to this definition"></a></dt>
<dd><p>Make a duration from a number of seconds.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Duration.from_usecs">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_usecs</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">usecs</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Duration.from_usecs" title="Link to this definition"></a></dt>
<dd><p>Make a duration from a number of microseconds.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Duration.to_msecs">
<span class="sig-name descname"><span class="pre">to_msecs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Duration.to_msecs" title="Link to this definition"></a></dt>
<dd><p>Convert to an integer number of whole milliseconds.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Duration.to_secs">
<span class="sig-name descname"><span class="pre">to_secs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Duration.to_secs" title="Link to this definition"></a></dt>
<dd><p>Convert to an integer number of whole seconds.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Duration.to_usecs">
<span class="sig-name descname"><span class="pre">to_usecs</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Duration.to_usecs" title="Link to this definition"></a></dt>
<dd><p>Convert to an integer number of whole microseconds.</p>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.Enum">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">Enum</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Enum" title="Link to this definition"></a></dt>
<dd><p>Generic enumeration.</p>
<p>Derive from this class to define new enumerations.</p>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.EventSet">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">EventSet</span></span><a class="headerlink" href="#isx.EventSet" title="Link to this definition"></a></dt>
<dd><p>An event set contains the event data of a number of components or cells.</p>
<p>It is typically derived from a cell set after applying an event detection
algorithm.
Each event of a cell is comprised of a time stamp offset and a value or amplitude.</p>
<section id="id16">
<h2>Examples<a class="headerlink" href="#id16" title="Link to this heading"></a></h2>
<p>Read an existing event set from a file and get the event data associated with the
first cell.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">event_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">EventSet</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP-BP-MC-DFF-PCA-ICA-ED.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">offsets</span><span class="p">,</span> <span class="n">amplitudes</span><span class="p">]</span> <span class="o">=</span> <span class="n">event_set</span><span class="o">.</span><span class="n">get_cell_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Write a new event set to a file by applying a threshold to the traces of an existing
cell set.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">cell_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">CellSet</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP-BP-MC-DFF-PCA-ICA.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cell_names</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;C</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">c</span><span class="p">)</span> <span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cell_set</span><span class="o">.</span><span class="n">num_cells</span><span class="p">)]</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">event_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">EventSet</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP-BP-MC-DFF-PCA-ICA-custom_ED.isxd&#39;</span><span class="p">,</span> <span class="n">cell_set</span><span class="o">.</span><span class="n">timing</span><span class="p">,</span> <span class="n">cell_names</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">offsets</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="n">x</span><span class="o">.</span><span class="n">to_usecs</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">cell_set</span><span class="o">.</span><span class="n">timing</span><span class="o">.</span><span class="n">get_offsets_since_start</span><span class="p">()],</span> <span class="n">numpy</span><span class="o">.</span><span class="n">uint64</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cell_set</span><span class="o">.</span><span class="n">num_cells</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">trace</span> <span class="o">=</span> <span class="n">cell_set</span><span class="o">.</span><span class="n">get_cell_trace_data</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">above_thresh</span> <span class="o">=</span> <span class="n">trace</span> <span class="o">&gt;</span> <span class="mi">500</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">event_set</span><span class="o">.</span><span class="n">set_cell_data</span><span class="p">(</span><span class="n">c</span><span class="p">,</span> <span class="n">offsets</span><span class="p">[</span><span class="n">above_thresh</span><span class="p">],</span> <span class="n">trace</span><span class="p">[</span><span class="n">above_thresh</span><span class="p">])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">event_set</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="id17">
<h2>Attributes<a class="headerlink" href="#id17" title="Link to this heading"></a></h2>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file that stores this.</p>
</dd>
<dt>mode<span class="classifier">{‘r’, ‘w’}</span></dt><dd><p>The mode the file was opened with.</p>
</dd>
<dt>timing<span class="classifier"><a class="reference internal" href="#isx.Timing" title="isx.Timing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Timing</span></code></a></span></dt><dd><p>The timing of the samples in each event trace.</p>
</dd>
<dt>num_cells<span class="classifier">int</span></dt><dd><p>The number of cells or components.</p>
</dd>
<dt>cell_dict<span class="classifier">dict</span></dt><dd><p>Dictionary mapping cell names to cell indices</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="isx.EventSet.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.EventSet.flush" title="Link to this definition"></a></dt>
<dd><p>Flush all meta-data and cell data to the file.</p>
<p>This should be called after setting all cell data of an event set opened with <a class="reference internal" href="#isx.EventSet.write" title="isx.EventSet.write"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.EventSet.write()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.EventSet.get_acquisition_info">
<span class="sig-name descname"><span class="pre">get_acquisition_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.EventSet.get_acquisition_info" title="Link to this definition"></a></dt>
<dd><p>Get information about acquisition that may be stored in some files,
such as nVista 3 movies and data derived from those.</p>
<section id="id18">
<h3>Returns<a class="headerlink" href="#id18" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>dict</dt><dd><p>A dictionary likely parsed from JSON that maps from string keys to variant values.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.EventSet.get_cell_data">
<span class="sig-name descname"><span class="pre">get_cell_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.EventSet.get_cell_data" title="Link to this definition"></a></dt>
<dd><p>Get the event data associated with a cell.</p>
<section id="id19">
<h3>Arguments<a class="headerlink" href="#id19" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
</dl>
</section>
<section id="id20">
<h3>Returns<a class="headerlink" href="#id20" title="Link to this heading"></a></h3>
<dl>
<dt>offsets<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The 1D array of time stamps offsets from the start in microseconds.</p>
</dd>
<dt>amplitudes<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The 1D array of event amplitudes.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.EventSet.get_cell_index">
<span class="sig-name descname"><span class="pre">get_cell_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.EventSet.get_cell_index" title="Link to this definition"></a></dt>
<dd><section id="id21">
<h3>Arguments<a class="headerlink" href="#id21" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>name<span class="classifier">int &gt;= 0</span></dt><dd><p>The name of a cell.</p>
</dd>
</dl>
</section>
<section id="id22">
<h3>Returns<a class="headerlink" href="#id22" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>str</dt><dd><p>The index of the named cell.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.EventSet.get_cell_name">
<span class="sig-name descname"><span class="pre">get_cell_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.EventSet.get_cell_name" title="Link to this definition"></a></dt>
<dd><section id="id23">
<h3>Arguments<a class="headerlink" href="#id23" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
</dl>
</section>
<section id="id24">
<h3>Returns<a class="headerlink" href="#id24" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>str</dt><dd><p>The name of the indexed cell.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.EventSet.read">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.EventSet.read" title="Link to this definition"></a></dt>
<dd><p>Open an existing event set from a file for reading.</p>
<p>This is a light weight operation that simply reads the meta-data from the event set,
and does not read any event data.</p>
<section id="id25">
<h3>Arguments<a class="headerlink" href="#id25" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to read.</p>
</dd>
</dl>
</section>
<section id="id26">
<h3>Returns<a class="headerlink" href="#id26" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.EventSet" title="isx.EventSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.EventSet</span></code></a></dt><dd><p>The event set that was read. Meta-data is immediately available.
Event data must be read using <a class="reference internal" href="#isx.EventSet.get_cell_data" title="isx.EventSet.get_cell_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.EventSet.get_cell_data()</span></code></a>.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.EventSet.set_cell_data">
<span class="sig-name descname"><span class="pre">set_cell_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">offsets</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">amplitudes</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.EventSet.set_cell_data" title="Link to this definition"></a></dt>
<dd><p>Set the event data of a cell.</p>
<section id="id27">
<h3>Arguments<a class="headerlink" href="#id27" title="Link to this heading"></a></h3>
<dl>
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a cell.</p>
</dd>
<dt>offsets<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The 1D array of time stamps offsets from the start in microseconds.</p>
</dd>
<dt>amplitudes<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The 1D array of event amplitudes.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.EventSet.write">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timing</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_names</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.EventSet.write" title="Link to this definition"></a></dt>
<dd><p>Open a new event set to a file for writing.</p>
<p>This is a light weight operation. It does not write any event data immediately.</p>
<section id="id28">
<h3>Arguments<a class="headerlink" href="#id28" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to write. If it already exists, this will error.</p>
</dd>
<dt>timing<span class="classifier">isx.Timing</span></dt><dd><p>The timing of the event set to write. Typically this comes from the cell set this
is derived from.</p>
</dd>
<dt>cell_names<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The names of the cells that will be written. Typically these come from the cell set
this is derived from.</p>
</dd>
</dl>
</section>
<section id="id29">
<h3>Returns<a class="headerlink" href="#id29" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.EventSet" title="isx.EventSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.EventSet</span></code></a></dt><dd><p>The empty event set that was written.
Image and trace data must be written with <a class="reference internal" href="#isx.EventSet.set_cell_data" title="isx.EventSet.set_cell_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.EventSet.set_cell_data()</span></code></a>.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.Fraction">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">Fraction</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">numerator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">denominator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">_normalize</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Fraction" title="Link to this definition"></a></dt>
<dd><p>This class implements rational numbers.</p>
<p>In the two-argument form of the constructor, Fraction(8, 6) will
produce a rational number equivalent to 4/3. Both arguments must
be Rational. The numerator defaults to 0 and the denominator
defaults to 1 so that Fraction(3) == 3 and Fraction() == 0.</p>
<p>Fractions can also be constructed from:</p>
<blockquote>
<div><ul class="simple">
<li><p>numeric strings similar to those accepted by the
float constructor (for example, ‘-2.3’ or ‘1e10’)</p></li>
<li><p>strings of the form ‘123/456’</p></li>
<li><p>float and Decimal instances</p></li>
<li><p>other Rational instances (including integers)</p></li>
</ul>
</div></blockquote>
<dl class="py method">
<dt class="sig sig-object py" id="isx.Fraction.as_integer_ratio">
<span class="sig-name descname"><span class="pre">as_integer_ratio</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Fraction.as_integer_ratio" title="Link to this definition"></a></dt>
<dd><p>Return the integer ratio as a tuple.</p>
<p>Return a tuple of two integers, whose ratio is equal to the
Fraction and with a positive denominator.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Fraction.from_decimal">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_decimal</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dec</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Fraction.from_decimal" title="Link to this definition"></a></dt>
<dd><p>Converts a finite Decimal instance to a rational number, exactly.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Fraction.from_float">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">from_float</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">f</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Fraction.from_float" title="Link to this definition"></a></dt>
<dd><p>Converts a finite float to a rational number, exactly.</p>
<p>Beware that Fraction.from_float(0.3) != Fraction(3, 10).</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Fraction.limit_denominator">
<span class="sig-name descname"><span class="pre">limit_denominator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">max_denominator</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1000000</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Fraction.limit_denominator" title="Link to this definition"></a></dt>
<dd><p>Closest Fraction to self with denominator at most max_denominator.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">Fraction</span><span class="p">(</span><span class="s1">&#39;3.141592653589793&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">limit_denominator</span><span class="p">(</span><span class="mi">10</span><span class="p">)</span>
<span class="go">Fraction(22, 7)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Fraction</span><span class="p">(</span><span class="s1">&#39;3.141592653589793&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">limit_denominator</span><span class="p">(</span><span class="mi">100</span><span class="p">)</span>
<span class="go">Fraction(311, 99)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">Fraction</span><span class="p">(</span><span class="mi">4321</span><span class="p">,</span> <span class="mi">8765</span><span class="p">)</span><span class="o">.</span><span class="n">limit_denominator</span><span class="p">(</span><span class="mi">10000</span><span class="p">)</span>
<span class="go">Fraction(4321, 8765)</span>
</pre></div>
</div>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.GpioSet">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">GpioSet</span></span><a class="headerlink" href="#isx.GpioSet" title="Link to this definition"></a></dt>
<dd><p>A GPIO set contains the data recorded across a number of channels.</p>
<p>Each data point is comprised of a time stamp offset and a value or amplitude.</p>
<section id="id30">
<h2>Examples<a class="headerlink" href="#id30" title="Link to this heading"></a></h2>
<p>Read an existing gpio set from a file and get the data associated with the first channel.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gpio_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">GpioSet</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;2020-05-20-10-33-22_video.gpio&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="p">[</span><span class="n">offsets</span><span class="p">,</span> <span class="n">amplitudes</span><span class="p">]</span> <span class="o">=</span> <span class="n">gpio_set</span><span class="o">.</span><span class="n">get_channel_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id31">
<h2>Attributes<a class="headerlink" href="#id31" title="Link to this heading"></a></h2>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file that stores this.</p>
</dd>
<dt>mode<span class="classifier">{‘r’, ‘w’}</span></dt><dd><p>The mode the file was opened with.</p>
</dd>
<dt>timing<span class="classifier"><a class="reference internal" href="#isx.Timing" title="isx.Timing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Timing</span></code></a></span></dt><dd><p>The timing of the samples.</p>
</dd>
<dt>num_channels<span class="classifier">int</span></dt><dd><p>The number of channels.</p>
</dd>
<dt>channel_dict<span class="classifier">dict</span></dt><dd><p>Dictionary mapping channel names to channel indices</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="isx.GpioSet.get_acquisition_info">
<span class="sig-name descname"><span class="pre">get_acquisition_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.GpioSet.get_acquisition_info" title="Link to this definition"></a></dt>
<dd><p>Get information about acquisition that may be stored in some files,
such as nVista 3 movies and data derived from those.</p>
<section id="id32">
<h3>Returns<a class="headerlink" href="#id32" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>dict</dt><dd><p>A dictionary likely parsed from JSON that maps from string keys to variant values.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.GpioSet.get_channel_data">
<span class="sig-name descname"><span class="pre">get_channel_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.GpioSet.get_channel_data" title="Link to this definition"></a></dt>
<dd><p>Get the data associated with a channel.</p>
<section id="id33">
<h3>Arguments<a class="headerlink" href="#id33" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a channel.</p>
</dd>
</dl>
</section>
<section id="id34">
<h3>Returns<a class="headerlink" href="#id34" title="Link to this heading"></a></h3>
<dl>
<dt>offsets<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The 1D array of time stamps offsets from the start in microseconds.</p>
</dd>
<dt>amplitudes<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The 1D array of amplitudes.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.GpioSet.get_channel_index">
<span class="sig-name descname"><span class="pre">get_channel_index</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">name</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.GpioSet.get_channel_index" title="Link to this definition"></a></dt>
<dd><section id="id35">
<h3>Arguments<a class="headerlink" href="#id35" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>name<span class="classifier">int &gt;= 0</span></dt><dd><p>The name of a channel.</p>
</dd>
</dl>
</section>
<section id="id36">
<h3>Returns<a class="headerlink" href="#id36" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>str</dt><dd><p>The index of the named channel.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.GpioSet.get_channel_name">
<span class="sig-name descname"><span class="pre">get_channel_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.GpioSet.get_channel_name" title="Link to this definition"></a></dt>
<dd><section id="id37">
<h3>Arguments<a class="headerlink" href="#id37" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a channel.</p>
</dd>
</dl>
</section>
<section id="id38">
<h3>Returns<a class="headerlink" href="#id38" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>str</dt><dd><p>The name of the indexed channel.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.GpioSet.read">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.GpioSet.read" title="Link to this definition"></a></dt>
<dd><p>Open an existing GPIO set from a file for reading.</p>
<p>This is a light weight operation that simply reads the meta-data from the GPIO set,
and does not read any GPIO data.</p>
<section id="id39">
<h3>Arguments<a class="headerlink" href="#id39" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to read.</p>
</dd>
</dl>
</section>
<section id="id40">
<h3>Returns<a class="headerlink" href="#id40" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.GpioSet" title="isx.GpioSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.GpioSet</span></code></a></dt><dd><p>The GPIO set that was read. Meta-data is immediately available.
GPIO data must be read using <a class="reference internal" href="#isx.GpioSet.get_channel_data" title="isx.GpioSet.get_channel_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.GpioSet.get_channel_data()</span></code></a>.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.Image">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">Image</span></span><a class="headerlink" href="#isx.Image" title="Link to this definition"></a></dt>
<dd><p>An image is effectively a movie with one frame and no timing.</p>
<p>It is always backed by a file, which can be read or written using this class.
See <span class="xref std std-ref">importMovie</span> for details on what formats are supported for read.
Only the native <cite>.isxd</cite> format is supported for write.</p>
<section id="id41">
<h2>Examples<a class="headerlink" href="#id41" title="Link to this heading"></a></h2>
<p>Read an existing image and get its data.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">image</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Image</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP-BP-Mean Image.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">image_data</span> <span class="o">=</span> <span class="n">image</span><span class="o">.</span><span class="n">get_data</span><span class="p">()</span>
</pre></div>
</div>
<p>Calculate the minimum image from an existing movie and write it.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">movie</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Movie</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">min_image</span> <span class="o">=</span> <span class="mi">4095</span> <span class="o">*</span> <span class="n">numpy</span><span class="o">.</span><span class="n">ones</span><span class="p">(</span><span class="n">movie</span><span class="o">.</span><span class="n">spacing</span><span class="o">.</span><span class="n">num_pixels</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="n">movie</span><span class="o">.</span><span class="n">data_type</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">movie</span><span class="o">.</span><span class="n">timing</span><span class="o">.</span><span class="n">num_samples</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">min_image</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">minimum</span><span class="p">(</span><span class="n">min_image</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">get_frame_data</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">isx</span><span class="o">.</span><span class="n">Image</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP-min.isxd&#39;</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">spacing</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">data_type</span><span class="p">,</span> <span class="n">min_image</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="id42">
<h2>Attributes<a class="headerlink" href="#id42" title="Link to this heading"></a></h2>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file that stores this.</p>
</dd>
<dt>mode<span class="classifier">{‘r’, ‘w’}</span></dt><dd><p>The mode the file was opened with.</p>
</dd>
<dt>spacing<span class="classifier"><a class="reference internal" href="#isx.Spacing" title="isx.Spacing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Spacing</span></code></a></span></dt><dd><p>The spacing of the pixels in the image.</p>
</dd>
<dt>data_type<span class="classifier">{numpy.uint16, numpy.float32}</span></dt><dd><p>The data type of each pixel.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="isx.Image.get_data">
<span class="sig-name descname"><span class="pre">get_data</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Image.get_data" title="Link to this definition"></a></dt>
<dd><p>Get the data stored in the image.</p>
<section id="id43">
<h3>Returns<a class="headerlink" href="#id43" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The image data.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Image.read">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Image.read" title="Link to this definition"></a></dt>
<dd><p>Read an existing image from a file.</p>
<section id="id44">
<h3>Arguments<a class="headerlink" href="#id44" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the image file to read.</p>
</dd>
</dl>
</section>
<section id="id45">
<h3>Returns<a class="headerlink" href="#id45" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.Image" title="isx.Image"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Image</span></code></a></dt><dd><p>The image that was read.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Image.write">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spacing</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Image.write" title="Link to this definition"></a></dt>
<dd><p>Write an image to a file.</p>
<section id="id46">
<h3>Arguments<a class="headerlink" href="#id46" title="Link to this heading"></a></h3>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to write. If it already exists, this will error.</p>
</dd>
<dt>spacing<span class="classifier"><a class="reference internal" href="#isx.Spacing" title="isx.Spacing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Spacing</span></code></a></span></dt><dd><p>The spacing of the image to write.</p>
</dd>
<dt>data_type<span class="classifier">{numpy.uint16, numpy.float32}</span></dt><dd><p>The data type of each pixel.</p>
</dd>
<dt>data<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.array</span></code></span></dt><dd><p>The 2D array of data to write.</p>
</dd>
</dl>
</section>
<section id="id47">
<h3>Returns<a class="headerlink" href="#id47" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.Image" title="isx.Image"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Image</span></code></a></dt><dd><p>The image that was written.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.Movie">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">Movie</span></span><a class="headerlink" href="#isx.Movie" title="Link to this definition"></a></dt>
<dd><p>A movie contains a number of frames with timing and spacing information.</p>
<p>It is always backed by a file, which can be read or written using this class.
See <span class="xref std std-ref">importMovie</span> for details on what formats are supported for read.
Only the native <cite>.isxd</cite> format is supported for write.</p>
<section id="id48">
<h2>Examples<a class="headerlink" href="#id48" title="Link to this heading"></a></h2>
<p>Read an existing movie and get its first frame as a numpy array.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">movie</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Movie</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">frame_data</span> <span class="o">=</span> <span class="n">movie</span><span class="o">.</span><span class="n">get_frame_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Write a 400x300 movie with 200 random frames of float32 values.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">timing</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Timing</span><span class="p">(</span><span class="n">num_samples</span><span class="o">=</span><span class="mi">200</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">spacing</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Spacing</span><span class="p">(</span><span class="n">num_pixels</span><span class="o">=</span><span class="p">(</span><span class="mi">300</span><span class="p">,</span> <span class="mi">400</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">movie</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Movie</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;movie-400x300x200.isxd&#39;</span><span class="p">,</span> <span class="n">timing</span><span class="p">,</span> <span class="n">spacing</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">timing</span><span class="o">.</span><span class="n">num_samples</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">movie</span><span class="o">.</span><span class="n">set_frame_data</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">spacing</span><span class="o">.</span><span class="n">num_pixels</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">movie</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="id49">
<h2>Attributes<a class="headerlink" href="#id49" title="Link to this heading"></a></h2>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file that stores this.</p>
</dd>
<dt>mode<span class="classifier">{‘r’, ‘w’}</span></dt><dd><p>The mode the file was opened with.</p>
</dd>
<dt>timing<span class="classifier"><a class="reference internal" href="#isx.Timing" title="isx.Timing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Timing</span></code></a></span></dt><dd><p>The timing of the frames.</p>
</dd>
<dt>spacing<span class="classifier"><a class="reference internal" href="#isx.Spacing" title="isx.Spacing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Spacing</span></code></a></span></dt><dd><p>The spacing of the pixels in each frame.</p>
</dd>
<dt>data_type<span class="classifier">{numpy.uint16, numpy.float32}</span></dt><dd><p>The data type of each pixel.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="isx.Movie.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Movie.flush" title="Link to this definition"></a></dt>
<dd><p>Flush all meta-data and frame data to the file.</p>
<p>This should be called after setting all frames of a movie opened with <a class="reference internal" href="#isx.Movie.write" title="isx.Movie.write"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.Movie.write()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Movie.get_acquisition_info">
<span class="sig-name descname"><span class="pre">get_acquisition_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Movie.get_acquisition_info" title="Link to this definition"></a></dt>
<dd><p>Get information about acquisition that may be stored in some files,
such as nVista 3 movies and data derived from those.</p>
<section id="id50">
<h3>Returns<a class="headerlink" href="#id50" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>dict</dt><dd><p>A dictionary likely parsed from JSON that maps from string keys to variant values.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Movie.get_frame_data">
<span class="sig-name descname"><span class="pre">get_frame_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Movie.get_frame_data" title="Link to this definition"></a></dt>
<dd><p>Get a frame from the movie by index.</p>
<section id="id51">
<h3>Arguments<a class="headerlink" href="#id51" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of the frame. If this is out of range, this should error.</p>
</dd>
</dl>
</section>
<section id="id52">
<h3>Returns<a class="headerlink" href="#id52" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The retrieved frame data.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Movie.get_frame_timestamp">
<span class="sig-name descname"><span class="pre">get_frame_timestamp</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Movie.get_frame_timestamp" title="Link to this definition"></a></dt>
<dd><p>Get a frame timestamp from the movie by index.</p>
<p>The timestamps are in units of microseconds.
This is a TSC (time stamp counter) value which is saved during acquisition.
These values come from a hardware counter on a particular acquisition box.
As a result, they can only be used to compare to other values that originate from the same hardware counter (e.g., paired recordings).</p>
<p>To get timestamps relative to the start of the movie,
simply subtract each timestamp with the timestamp of the first frame in the movie.
To get timestamps relative to Unix epoch time, add the timestamps computed relative
to the start of the movie with the Unix epoch start timestamp of the movie, accessible through the <cite>timing</cite> member of this class.
Alternatively, the timestamps of a movie can be exported relative to the start of the movie,
or relative to the Unix epoch time, using the function <cite>isx.export_movie_timestamps_to_csv</cite>.</p>
<section id="id53">
<h3>Arguments<a class="headerlink" href="#id53" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of the frame. If this is out of range, this should error.</p>
</dd>
</dl>
</section>
<section id="id54">
<h3>Returns<a class="headerlink" href="#id54" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>int</dt><dd><p>The retreived frame timestamp.
If the movie has no frame timestamps, the function will throw an error,</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Movie.read">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Movie.read" title="Link to this definition"></a></dt>
<dd><p>Open an existing movie from a file for reading.</p>
<p>This is a light weight operation that simply reads the meta-data from the movie,
and does not read any frame data.</p>
<section id="id55">
<h3>Arguments<a class="headerlink" href="#id55" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to read.</p>
</dd>
</dl>
</section>
<section id="id56">
<h3>Returns<a class="headerlink" href="#id56" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.Movie" title="isx.Movie"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Movie</span></code></a></dt><dd><p>The movie that was read. Meta-data is immediately available.
Frames must be read using <code class="xref py py-func docutils literal notranslate"><span class="pre">isx.Movie.get_frame()</span></code>.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Movie.set_frame_data">
<span class="sig-name descname"><span class="pre">set_frame_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Movie.set_frame_data" title="Link to this definition"></a></dt>
<dd><p>Set frame data in a writable movie.</p>
<p>Frames must be set in increasing order, otherwise this will error.</p>
<section id="id57">
<h3>Arguments<a class="headerlink" href="#id57" title="Link to this heading"></a></h3>
<dl>
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of the frame.</p>
</dd>
<dt>frame<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The frame data.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Movie.write">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timing</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spacing</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">data_type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Movie.write" title="Link to this definition"></a></dt>
<dd><p>Open a new movie to a file for writing.</p>
<p>This is a light weight operation. It does not write any frame data immediately.</p>
<section id="id58">
<h3>Arguments<a class="headerlink" href="#id58" title="Link to this heading"></a></h3>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to write. If it already exists, this will error.</p>
</dd>
<dt>timing<span class="classifier"><a class="reference internal" href="#isx.Timing" title="isx.Timing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Timing</span></code></a></span></dt><dd><p>The timing of the movie to write.</p>
</dd>
<dt>spacing<span class="classifier"><a class="reference internal" href="#isx.Spacing" title="isx.Spacing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Spacing</span></code></a></span></dt><dd><p>The spacing of the movie to write.</p>
</dd>
<dt>data_type<span class="classifier">{numpy.uint16, numpy.float32}</span></dt><dd><p>The data type of each pixel.</p>
</dd>
</dl>
</section>
<section id="id59">
<h3>Returns<a class="headerlink" href="#id59" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.Movie" title="isx.Movie"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Movie</span></code></a></dt><dd><p>The empty movie that was written.
Frame data must be written with <a class="reference internal" href="#isx.Movie.set_frame_data" title="isx.Movie.set_frame_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.Movie.set_frame_data()</span></code></a>.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.Spacing">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">Spacing</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_pixels</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">(0,</span> <span class="pre">0)</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Spacing" title="Link to this definition"></a></dt>
<dd><p>The spacing associated with a set of pixels.</p>
<section id="id60">
<h2>Examples<a class="headerlink" href="#id60" title="Link to this heading"></a></h2>
<p>Make spacing for a 1440x1080 image.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">spacing</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Spacing</span><span class="p">(</span><span class="n">num_pixels</span><span class="o">=</span><span class="p">(</span><span class="mi">1080</span><span class="p">,</span> <span class="mi">1440</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="id61">
<h2>Attributes<a class="headerlink" href="#id61" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>num_pixels<span class="classifier">2-tuple&lt;int&gt;</span></dt><dd><p>The number of pixels as (num_rows, num_cols).</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.Time">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">Time</span></span><a class="headerlink" href="#isx.Time" title="Link to this definition"></a></dt>
<dd><p>A time stamp that defines a calendar date and wall time.</p>
<dl class="py method">
<dt class="sig sig-object py" id="isx.Time.to_datetime">
<span class="sig-name descname"><span class="pre">to_datetime</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Time.to_datetime" title="Link to this definition"></a></dt>
<dd><section id="id62">
<h2>Returns<a class="headerlink" href="#id62" title="Link to this heading"></a></h2>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">datetime.datetime</span></code></dt><dd><p>The nearest Python datetime.</p>
</dd>
</dl>
</section>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.Timing">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">Timing</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">num_samples=0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">period=&lt;isx.core.Duration</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">start=&lt;isx.core.Time</span> <span class="pre">object&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dropped=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cropped=[]</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">blank=[]</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.Timing" title="Link to this definition"></a></dt>
<dd><p>The timing associated with a set of samples, such as the frames of a movie or the
values of a trace.</p>
<p>Some samples are described as invalid, meaning that the sample is missing.
These include dropped samples, which could arise due to an error at acquisition time,
and cropped samples, which are likely due to processing.</p>
<section id="id63">
<h2>Examples<a class="headerlink" href="#id63" title="Link to this heading"></a></h2>
<p>Make timing for 9 samples, every 10 milliseconds.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">timing</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Timing</span><span class="p">(</span><span class="n">num_samples</span><span class="o">=</span><span class="mi">9</span><span class="p">,</span> <span class="n">period</span><span class="o">=</span><span class="n">isx</span><span class="o">.</span><span class="n">Duration</span><span class="o">.</span><span class="n">from_msecs</span><span class="p">(</span><span class="mi">10</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="id64">
<h2>Attributes<a class="headerlink" href="#id64" title="Link to this heading"></a></h2>
<dl>
<dt>num_samples<span class="classifier">int &gt;= 0</span></dt><dd><p>The number of samples, including invalid (dropped, cropped, or blank) ones.</p>
</dd>
<dt>period<span class="classifier"><a class="reference internal" href="#isx.Duration" title="isx.Duration"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Duration</span></code></a></span></dt><dd><p>The period or duration of one sample.</p>
</dd>
<dt>start<span class="classifier"><a class="reference internal" href="#isx.Time" title="isx.Time"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Time</span></code></a></span></dt><dd><p>The time stamp associated with the start of the first sample.</p>
</dd>
<dt>dropped<span class="classifier">list&lt;int&gt;</span></dt><dd><p>The indices of the dropped samples.</p>
</dd>
<dt>cropped<span class="classifier">list&lt;2-tuple&gt;</span></dt><dd><p>The index ranges of the cropped samples.
Each element specifies the inclusive lower and upper bounds of a range
of indices.</p>
</dd>
<dt>blank<span class="classifier">list&lt;int&gt;</span></dt><dd><p>The indices of the blank samples.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="isx.Timing.get_offsets_since_start">
<span class="sig-name descname"><span class="pre">get_offsets_since_start</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Timing.get_offsets_since_start" title="Link to this definition"></a></dt>
<dd><p>Get the offsets from the start of the timing.</p>
<section id="id65">
<h3>Returns<a class="headerlink" href="#id65" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>list&lt;<a class="reference internal" href="#isx.Duration" title="isx.Duration"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Duration</span></code></a>&gt;</dt><dd><p>Each element is the offset from the start to a sample.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Timing.get_valid_samples">
<span class="sig-name descname"><span class="pre">get_valid_samples</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Timing.get_valid_samples" title="Link to this definition"></a></dt>
<dd><section id="id66">
<h3>Returns<a class="headerlink" href="#id66" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The valid sample indices.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.Timing.get_valid_samples_mask">
<span class="sig-name descname"><span class="pre">get_valid_samples_mask</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.Timing.get_valid_samples_mask" title="Link to this definition"></a></dt>
<dd><p>Get a 1D array mask indicating whether each sample is valid.</p>
<section id="id67">
<h3>Returns<a class="headerlink" href="#id67" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>Each element indicates whether the corresponding sample is valid.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
</dd></dl>

<dl class="py class">
<dt class="sig sig-object py" id="isx.VesselSet">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">VesselSet</span></span><a class="headerlink" href="#isx.VesselSet" title="Link to this definition"></a></dt>
<dd><p>A vessel set contains the image, line and trace data associated with components in
a movie, such as vessels or regions of interest.</p>
<p>It is always backed by a file in the native <cite>.isxd</cite> format.</p>
<p>A vessel set can represent two types of data: vessel diameter and rbc velocity.
Depending on the vessel type, different information will be stored to disk.</p>
<p>Note: Since blood flow algorithms apply a sliding window over input movies,
the timing of a vessel trace is different from the timing of its input movie.
Each frame of a vessel trace represents a measurement for a particular window sampled from its input movie.
Relative to the input movie, each frame maps to the start of the corresponding window sampled.
The duration of a frame is equal to the time increment of the sliding window.</p>
<p>The following examples will demonstrate the types of data accessible for both types of vessel sets.</p>
<section id="id68">
<h2>Examples<a class="headerlink" href="#id68" title="Link to this heading"></a></h2>
<p><strong>Vessel Diameter</strong>
Read an existing vessel set from a file and get the image, line and diameter trace data of
the first vessel.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vessel_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">VesselSet</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;bloodflow_movie_10s-VD.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">image_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_image_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">line_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_line_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">trace_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_trace_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">center_trace_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_center_trace_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
<p>Write a new vessel set to a file with the same timing and spacing as an
existing movie, with 3 random vessel images, lines and diameter traces.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">movie</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Movie</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vessel_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">VesselSet</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;vessel_set.isxd&#39;</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">timing</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">spacing</span><span class="p">,</span> <span class="s1">&#39;vessel diameter&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">image</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">vessel_set</span><span class="o">.</span><span class="n">spacing</span><span class="o">.</span><span class="n">num_pixels</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">lines</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">spacing</span><span class="o">.</span><span class="n">num_pixels</span><span class="p">),</span> <span class="p">(</span><span class="n">num_vessels</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">trace</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">vessel_set</span><span class="o">.</span><span class="n">timing</span><span class="o">.</span><span class="n">num_samples</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">center_trace</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">vessel_set</span><span class="o">.</span><span class="n">timing</span><span class="o">.</span><span class="n">num_samples</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">vessel_set</span><span class="o">.</span><span class="n">set_vessel_diameter_data</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">image</span><span class="p">,</span> <span class="n">lines</span><span class="p">,</span> <span class="n">trace</span><span class="p">,</span> <span class="n">center_trace</span><span class="p">,</span> <span class="s1">&#39;V</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vessel_set</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
</pre></div>
</div>
<p><strong>RBC Velocity</strong>
Read an existing vessel set from a file and get the image, line and rbc velocity trace data of
the first vessel.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">vessel_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">VesselSet</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;bloodflow_movie_10s-RBCV.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">image_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_image_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">line_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_line_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">trace_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_trace_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">direction_trace_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_direction_trace_data</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">corr_0</span> <span class="o">=</span> <span class="n">vessel_set</span><span class="o">.</span><span class="n">get_vessel_correlations_data</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">)</span> <span class="c1"># look at first frame of correlation data for the first vessel</span>
</pre></div>
</div>
<p>Write a new vessel set to a file with the same timing and spacing as an
existing movie, with 3 random vessel images, lines and rbc velocity traces.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">movie</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">Movie</span><span class="o">.</span><span class="n">read</span><span class="p">(</span><span class="s1">&#39;recording_20160613_105808-PP-PP.isxd&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vessel_set</span> <span class="o">=</span> <span class="n">isx</span><span class="o">.</span><span class="n">VesselSet</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="s1">&#39;vessel_set.isxd&#39;</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">timing</span><span class="p">,</span> <span class="n">movie</span><span class="o">.</span><span class="n">spacing</span><span class="p">,</span> <span class="s1">&#39;rbc velocity&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">3</span><span class="p">):</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">image</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">vessel_set</span><span class="o">.</span><span class="n">spacing</span><span class="o">.</span><span class="n">num_pixels</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">lines</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">min</span><span class="p">(</span><span class="n">spacing</span><span class="o">.</span><span class="n">num_pixels</span><span class="p">),</span> <span class="p">(</span><span class="n">num_vessels</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">trace</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">vessel_set</span><span class="o">.</span><span class="n">timing</span><span class="o">.</span><span class="n">num_samples</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">direction_trace</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">(</span><span class="n">vessel_set</span><span class="o">.</span><span class="n">timing</span><span class="o">.</span><span class="n">num_samples</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">correlation_size</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">2</span><span class="p">,</span> <span class="mi">5</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="p">(</span><span class="mi">2</span><span class="p">,))</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">correlations_trace</span> <span class="o">=</span> <span class="n">numpy</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">random</span><span class="p">([</span><span class="n">vessel_set</span><span class="o">.</span><span class="n">timing</span><span class="o">.</span><span class="n">num_samples</span><span class="p">,</span> <span class="mi">3</span><span class="p">,</span> <span class="n">correlation_size</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">correlation_size</span><span class="p">[</span><span class="mi">1</span><span class="p">]])</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="n">numpy</span><span class="o">.</span><span class="n">float32</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">vessel_set</span><span class="o">.</span><span class="n">set_rbc_velocity_data</span><span class="p">(</span><span class="n">i</span><span class="p">,</span> <span class="n">image</span><span class="p">,</span> <span class="n">lines</span><span class="p">,</span> <span class="n">trace</span><span class="p">,</span> <span class="n">direction_trace</span><span class="p">,</span> <span class="n">correlations_trace</span><span class="p">,</span> <span class="s1">&#39;V</span><span class="si">{}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">i</span><span class="p">))</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">vessel_set</span><span class="o">.</span><span class="n">flush</span><span class="p">()</span>
</pre></div>
</div>
</section>
<section id="id69">
<h2>Attributes<a class="headerlink" href="#id69" title="Link to this heading"></a></h2>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file that stores this.</p>
</dd>
<dt>mode<span class="classifier">{‘r’, ‘w’}</span></dt><dd><p>The mode the file was opened with.</p>
</dd>
<dt>timing<span class="classifier"><a class="reference internal" href="#isx.Timing" title="isx.Timing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Timing</span></code></a></span></dt><dd><p>The timing of the samples in each vessel trace.</p>
</dd>
<dt>spacing<span class="classifier"><a class="reference internal" href="#isx.Spacing" title="isx.Spacing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Spacing</span></code></a></span></dt><dd><p>The spacing of the pixels in each vessel image.</p>
</dd>
<dt>num_vessels<span class="classifier">int</span></dt><dd><p>The number of vessels or components.</p>
</dd>
</dl>
<dl class="py class">
<dt class="sig sig-object py" id="isx.VesselSet.VesselSetType">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">VesselSetType</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">value</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.VesselSetType" title="Link to this definition"></a></dt>
<dd><p>An enumeration.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.flush">
<span class="sig-name descname"><span class="pre">flush</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.flush" title="Link to this definition"></a></dt>
<dd><p>Flush all meta-data and vessel data to the file.</p>
<p>This should be called after setting all vessel data of a vessel set opened with <a class="reference internal" href="#isx.VesselSet.write" title="isx.VesselSet.write"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.VesselSet.write()</span></code></a>.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_acquisition_info">
<span class="sig-name descname"><span class="pre">get_acquisition_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_acquisition_info" title="Link to this definition"></a></dt>
<dd><p>Get information about acquisition that may be stored in some files,
such as nVista 3 movies and data derived from those.</p>
<section id="id70">
<h3>Returns<a class="headerlink" href="#id70" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>dict</dt><dd><p>A dictionary likely parsed from JSON that maps from string keys to variant values.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_center_trace_data">
<span class="sig-name descname"><span class="pre">get_vessel_center_trace_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_center_trace_data" title="Link to this definition"></a></dt>
<dd><p>Get the center trace data associated with a vessel.
This represents an index on the user drawn line where the center of the diamter was estimated to be.</p>
<section id="id71">
<h3>Arguments<a class="headerlink" href="#id71" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
</dl>
</section>
<section id="id72">
<h3>Returns<a class="headerlink" href="#id72" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The center trace data in a 1D array.
If no center traces are stored in the file, the function will throw an error</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_correlations_data">
<span class="sig-name descname"><span class="pre">get_vessel_correlations_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_correlations_data" title="Link to this definition"></a></dt>
<dd><p>Get the correlation trace data associated with a vessel at a certain frame.</p>
<section id="id73">
<h3>Arguments<a class="headerlink" href="#id73" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
<dt>frame<span class="classifier">int &gt;= 0</span></dt><dd><p>The frame index of the time-series trace.</p>
</dd>
</dl>
</section>
<section id="id74">
<h3>Returns<a class="headerlink" href="#id74" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The correlations data in a 3D array of size (3, width, height)
Each slice of the ndarray is a correlation heatmap for a particular temporal offset
The mapping of temporal offsets is (slice 0 -&gt; t = -1, slice 1 -&gt; t = 0, slice 2 = t = 1)
If no correlations are stored in the file, the function will throw an error</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_direction_trace_data">
<span class="sig-name descname"><span class="pre">get_vessel_direction_trace_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_direction_trace_data" title="Link to this definition"></a></dt>
<dd><p>Get the direction trace data associated with a vessel
This is the direction component of each velocity measurement reported in degrees relative to positive x-axis.</p>
<section id="id75">
<h3>Arguments<a class="headerlink" href="#id75" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
</dl>
</section>
<section id="id76">
<h3>Returns<a class="headerlink" href="#id76" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The direction trace data in a 1D array.
If no direction traces are stored in the file, the function will throw an error</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_image_data">
<span class="sig-name descname"><span class="pre">get_vessel_image_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_image_data" title="Link to this definition"></a></dt>
<dd><p>Get the image data associated with a vessel.</p>
<section id="id77">
<h3>Arguments<a class="headerlink" href="#id77" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
</dl>
</section>
<section id="id78">
<h3>Returns<a class="headerlink" href="#id78" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The image data in a 2D array.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_line_data">
<span class="sig-name descname"><span class="pre">get_vessel_line_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_line_data" title="Link to this definition"></a></dt>
<dd><p>Get the line data associated with a vessel.</p>
<section id="id79">
<h3>Arguments<a class="headerlink" href="#id79" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
</dl>
</section>
<section id="id80">
<h3>Returns<a class="headerlink" href="#id80" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The line data in a 2D array.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_name">
<span class="sig-name descname"><span class="pre">get_vessel_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_name" title="Link to this definition"></a></dt>
<dd><section id="id81">
<h3>Arguments<a class="headerlink" href="#id81" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
</dl>
</section>
<section id="id82">
<h3>Returns<a class="headerlink" href="#id82" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>str</dt><dd><p>The name of the indexed vessel.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_set_type">
<span class="sig-name descname"><span class="pre">get_vessel_set_type</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_set_type" title="Link to this definition"></a></dt>
<dd><p>Get the type of the vessel set.</p>
<section id="id83">
<h3>Returns<a class="headerlink" href="#id83" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>VesselSetType Enum</dt><dd><p>An enum representation of the different vessel set types.
either VesselSetType.VESSEL_DIAMETER or VesselSetType.RBC_VELOCITY</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_status">
<span class="sig-name descname"><span class="pre">get_vessel_status</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_status" title="Link to this definition"></a></dt>
<dd><section id="id84">
<h3>Arguments<a class="headerlink" href="#id84" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
</dl>
</section>
<section id="id85">
<h3>Returns<a class="headerlink" href="#id85" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>{‘accepted’, ‘undecided’, ‘rejected’}</dt><dd><p>The status of the indexed vessel as a string.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.get_vessel_trace_data">
<span class="sig-name descname"><span class="pre">get_vessel_trace_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.get_vessel_trace_data" title="Link to this definition"></a></dt>
<dd><p>Get the trace data associated with a vessel.</p>
<section id="id86">
<h3>Arguments<a class="headerlink" href="#id86" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
</dl>
</section>
<section id="id87">
<h3>Returns<a class="headerlink" href="#id87" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></dt><dd><p>The trace data in a 1D array.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.has_correlation_heatmaps">
<span class="sig-name descname"><span class="pre">has_correlation_heatmaps</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.has_correlation_heatmaps" title="Link to this definition"></a></dt>
<dd><p>If true, cross-correlation heatmaps for rbc velocity measurements are stored in this file.</p>
<section id="id88">
<h3>Returns<a class="headerlink" href="#id88" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>bool</dt><dd><p>Flag indicating whether heatmaps were saved.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.read">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_only</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.read" title="Link to this definition"></a></dt>
<dd><p>Open an existing vessel set from a file for reading.</p>
<p>This is a light weight operation that simply reads the meta-data from the vessel set,
and does not read any image or trace data.</p>
<section id="id89">
<h3>Arguments<a class="headerlink" href="#id89" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to read.</p>
</dd>
<dt>read_only<span class="classifier">bool</span></dt><dd><p>If true, only allow meta-data and data to be read, otherwise allow some meta-data
to be written (e.g. vessel status).</p>
</dd>
</dl>
</section>
<section id="id90">
<h3>Returns<a class="headerlink" href="#id90" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.VesselSet" title="isx.VesselSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.VesselSet</span></code></a></dt><dd><p>The vessel set that was read. Meta-data is immediately available.
Image and trace data must be read using <a class="reference internal" href="#isx.VesselSet.get_vessel_image_data" title="isx.VesselSet.get_vessel_image_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.VesselSet.get_vessel_image_data()</span></code></a>
and <a class="reference internal" href="#isx.VesselSet.get_vessel_trace_data" title="isx.VesselSet.get_vessel_trace_data"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.VesselSet.get_vessel_trace_data()</span></code></a> respectively.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.set_rbc_velocity_data">
<span class="sig-name descname"><span class="pre">set_rbc_velocity_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">line</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">direction_trace</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">correlations_trace</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.set_rbc_velocity_data" title="Link to this definition"></a></dt>
<dd><p>Set the image, line and rbc velocity trace data of a vessel.</p>
<p>Vessels must be set in increasing order, otherwise this will error.</p>
<section id="id91">
<h3>Arguments<a class="headerlink" href="#id91" title="Link to this heading"></a></h3>
<dl>
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
<dt>image<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The image data in a 2D array.</p>
</dd>
<dt>line<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The line endpoint data in a 2D array.</p>
</dd>
<dt>trace<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The trace data in a 1D array.</p>
</dd>
<dt>direction_trace<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The direction trace data in a 1D array.</p>
</dd>
<dt>correlations_trace<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The correlations trace data in a 4D array T x 3 x W x H
Where T is the number of time samples
W is the width of the correlation heatmap
H is the width of the correlation heatmap
There are three heatmaps for each time sample representing the three temporal offsets (-1, 0, 1)</p>
</dd>
<dt>name<span class="classifier">str</span></dt><dd><p>The name of the vessel.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.set_vessel_diameter_data">
<span class="sig-name descname"><span class="pre">set_vessel_diameter_data</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">image</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">line</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">trace</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">center_trace</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">name</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.set_vessel_diameter_data" title="Link to this definition"></a></dt>
<dd><p>Set the image, line and diameter trace data of a vessel</p>
<p>Vessels must be set in increasing order, otherwise this will error.</p>
<section id="id92">
<h3>Arguments<a class="headerlink" href="#id92" title="Link to this heading"></a></h3>
<dl>
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
<dt>image<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The image data in a 2D array.</p>
</dd>
<dt>line<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The line endpoint data in a 2D array.</p>
</dd>
<dt>trace<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The trace data in a 1D array.</p>
</dd>
<dt>center_trace<span class="classifier"><code class="xref py py-class docutils literal notranslate"><span class="pre">numpy.ndarray</span></code></span></dt><dd><p>The center trace data in a 1D array.</p>
</dd>
<dt>name<span class="classifier">str</span></dt><dd><p>The name of the vessel.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.set_vessel_status">
<span class="sig-name descname"><span class="pre">set_vessel_status</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">index</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">status</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.set_vessel_status" title="Link to this definition"></a></dt>
<dd><p>Set the status of vessel. This will also flush the file.</p>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>As this flushes the file, only use this after all vessels have been
written using <code class="xref py py-func docutils literal notranslate"><span class="pre">isx.VesselSet.set_vessel_data()</span></code>.</p>
</div>
<section id="id93">
<h3>Arguments<a class="headerlink" href="#id93" title="Link to this heading"></a></h3>
<dl class="simple">
<dt>index<span class="classifier">int &gt;= 0</span></dt><dd><p>The index of a vessel.</p>
</dd>
<dt>status<span class="classifier">{‘accepted’, ‘undecided’, ‘rejected’}</span></dt><dd><p>The desired status of the indexed vessel.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="isx.VesselSet.write">
<em class="property"><span class="pre">classmethod</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">write</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">timing</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">spacing</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">vessel_type</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.VesselSet.write" title="Link to this definition"></a></dt>
<dd><p>Open a new vessel set to a file for writing.</p>
<p>This is a light weight operation. It does not write any image or trace data immediately.</p>
<section id="id94">
<h3>Arguments<a class="headerlink" href="#id94" title="Link to this heading"></a></h3>
<dl>
<dt>file_path<span class="classifier">str</span></dt><dd><p>The path of the file to write. If it already exists, this will error.</p>
</dd>
<dt>timing<span class="classifier"><a class="reference internal" href="#isx.Timing" title="isx.Timing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Timing</span></code></a></span></dt><dd><p>The timing of the vessel set to write. Typically this comes from the movie this
is derived from.</p>
</dd>
<dt>spacing<span class="classifier"><a class="reference internal" href="#isx.Spacing" title="isx.Spacing"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.Spacing</span></code></a></span></dt><dd><p>The spacing of the movie to write. Typically this comes from the movie this is
derived from.</p>
</dd>
<dt>vessel_type<span class="classifier">str</span></dt><dd><p>The type of metric to store in the vessel set. Either ‘vessel diameter’ or ‘rbc velocity’.</p>
</dd>
</dl>
</section>
<section id="id95">
<h3>Returns<a class="headerlink" href="#id95" title="Link to this heading"></a></h3>
<dl class="simple">
<dt><a class="reference internal" href="#isx.VesselSet" title="isx.VesselSet"><code class="xref py py-class docutils literal notranslate"><span class="pre">isx.VesselSet</span></code></a></dt><dd><p>The empty vessel set that was written.
Image and trace data must be written with <code class="xref py py-func docutils literal notranslate"><span class="pre">isx.VesselSet.set_vessel_data()</span></code>.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.align_start_times">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">align_start_times</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_ref_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_align_files</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.align_start_times" title="Link to this definition"></a></dt>
<dd><p>Align the epoch start times of files originating from the same paired and synchronized start-stop recording session.
The epoch start time stored in the input align files are modified in-place
so that they are aligned relative to the epoch start time of the input timing reference file</p>
<dl class="simple">
<dt>For each input align file, the epoch start time is recomputed using the following formula:</dt><dd><p>align_epoch_start_ms = ref_epoch_start_ms + ((align_first_tsc_us - ref_first_tsc_us) / 1e3)</p>
</dd>
<dt>In the event that the first sample of an input align file is dropped, the tsc value of the first sample is inferred using the following formula:</dt><dd><p>align_first_tsc_us = align_first_valid_tsc_us - (align_first_valid_idx * align_sample_period_us)</p>
</dd>
</dl>
<section id="id96">
<h2>Arguments<a class="headerlink" href="#id96" title="Link to this heading"></a></h2>
<blockquote>
<div><dl class="simple">
<dt>input_ref_file<span class="classifier">str</span></dt><dd><p>The path of the file to use as the timing reference to align with the other input files.
This can be either a .gpio file, .isxd movie, or .isxb movie, otherwise the function will throw an error.
If the timing reference is a movie, the movie must contain frame timestamps, otherwise this function will throw an error.</p>
</dd>
<dt>input_align_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The path of the files to align to the epoch start time of the input timing reference file.
These files can either be an .isxd movie or .isxb movie, otherwise the function will throw an error.
The movies must contain frame timestamps, otherwise this function will throw an error.</p>
</dd>
</dl>
</div></blockquote>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.convert_type_numpy_array">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">convert_type_numpy_array</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">array</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">dtype=&lt;class</span> <span class="pre">'numpy.uint16'&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">keep_0_to_1=False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.convert_type_numpy_array" title="Link to this definition"></a></dt>
<dd><p>Convert a numpy array to a different data type by normalizing and mapping.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.create_cell_map">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">create_cell_map</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">isxd_cellset_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">selected_cell_statuses</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">['accepted']</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">binary</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rgb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_isxd_cell_map_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_tiff_cell_map_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.create_cell_map" title="Link to this definition"></a></dt>
<dd><p>Generate cell map from an .isxd cellset file, saving the image as isxd and tiff.</p>
<section id="id97">
<h2>Arguments<a class="headerlink" href="#id97" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>isxd_cellset_file<span class="classifier">str</span></dt><dd><p>Path to an .isxd cellset file.</p>
</dd>
<dt>selected_cell_statuses<span class="classifier">list&lt;str&gt;</span></dt><dd><p>a list of cell statuses for decision criteria to keep cell footprints.
Possible list values are ‘accepted’, ‘undecided’, ‘rejected’.</p>
</dd>
<dt>cell_thresh<span class="classifier">float [0, 1]</span></dt><dd><p>Pixels with values lower than cell_thresh will be set to 0.</p>
</dd>
<dt>binary<span class="classifier">Bool</span></dt><dd><p>If true, pixels with values above cell_thresh are set to 1. (Pixels with values
below cell_thresh are set to 0.)</p>
</dd>
<dt>rgb<span class="classifier">one of “red”, “blue”, or “green”</span></dt><dd><p>Color for the cell map.</p>
</dd>
<dt>output_isxd_cell_map_file<span class="classifier">str</span></dt><dd><p>Path to the output isxd cell map image. If not given, will not be generated.</p>
</dd>
<dt>output_tiff_cell_map_file<span class="classifier">str</span></dt><dd><p>Path to the output tiff cell map image. If not given, will not be generated.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_aligned_timestamps">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_aligned_timestamps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_ref_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_align_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_ref_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_align_names</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_csv_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_aligned_timestamps" title="Link to this definition"></a></dt>
<dd><p>Export timestamps from files which originate from the same paired and synced start-stop recording session to a .csv file.
Timestamps are aligned to a single start time which is defined as the start time of the specified timing reference file.</p>
<section id="id98">
<h2>Arguments<a class="headerlink" href="#id98" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_ref_file<span class="classifier">str</span></dt><dd><p>The path of the file to use as the timing reference to align with the other input files.
Timestamps are exported relative to the start time of this file.
This can be either a .gpio file, .isxd movie, or .isxb movie, otherwise the function will throw an error.
If the timing reference is a movie, the movie must contain frame timestamps, otherwise this function will throw an error.</p>
</dd>
<dt>input_align_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The path of the files to align to the epoch start time of the input timing reference file.
These files can either be a .gpio file, .isxd movie, or .isxb movie, otherwise the function will throw an error.
The movies must contain frame timestamps, otherwise this function will throw an error.</p>
</dd>
<dt>input_ref_name<span class="classifier">str</span></dt><dd><p>The name of the reference data set to use in the output csv.</p>
</dd>
<dt>input_align_names<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The names of the align data sets to use in the output csv.</p>
</dd>
<dt>output_csv_file<span class="classifier">str</span></dt><dd><p>The path of the csv file to be written.</p>
</dd>
<dt>time_ref<span class="classifier">{‘start’, ‘unix’, ‘tsc’}</span></dt><dd><p>The time reference for the CSV time stamps.
If ‘start’ is used, the timestamps represent the seconds since the start of the movie.
If ‘unix’ is used, the timestamps represent the seconds since the Unix epoch.
If ‘tsc’ is used, the timestamps represent the hardware counter value on the acquisition box when each frame was captured.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_cell_set_to_csv_tiff">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_cell_set_to_csv_tiff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_cell_set_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_csv_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_tiff_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_props_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_cell_set_to_csv_tiff" title="Link to this definition"></a></dt>
<dd><p>Export cell sets to a CSV file with trace data and TIFF files with image data.</p>
<p>For more details see <span class="xref std std-ref">exportCellsAndStuff</span>.</p>
<p>Unlike the desktop application, this will only produce a TIFF cell map image file
and not a PNG file too.</p>
<section id="id99">
<h2>Arguments<a class="headerlink" href="#id99" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_cell_set_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the cell sets to export.</p>
</dd>
<dt>output_csv_file<span class="classifier">str</span></dt><dd><p>The path of the CSV file to write.</p>
</dd>
<dt>output_tiff_file<span class="classifier">str</span></dt><dd><p>The base name of the TIFF files to write.</p>
</dd>
<dt>time_ref<span class="classifier">{‘start’, ‘unix’}</span></dt><dd><p>The time reference for the CSV time stamps.
If ‘start’ is used, the time stamps represent the seconds since the start of the cell set.
If ‘unix’ is used, the time stamps represents the second since the Unix epoch.</p>
</dd>
<dt>output_props_file<span class="classifier">str</span></dt><dd><p>The path of the properties CSV file to write.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_ethovision_data_with_isxb_timestamps">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_ethovision_data_with_isxb_timestamps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_ethovision_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_isxb_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_csv_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_ref_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_ethovision_data_with_isxb_timestamps" title="Link to this definition"></a></dt>
<dd><dl class="simple">
<dt>Given paths to an Ethovision file and a nVision (.isxb) movie file,</dt><dd><p>writes a csv file with Ethovision data and a dedicated column,
either <cite>isxb Frame Timestamp (s)</cite> or <cite>isxb Frame Timestamp (us)</cite>
depending on the format of the timestamps,
that contains aligned timestamps from the nVision movie.</p>
</dd>
<dt>It is important to note that internal Inscopix testing has shown that the Ethovision output file</dt><dd><p>can have a length that is one less than the number of frames in the input movie. When this is true,
the missing data row appears to be at the beginning.
This function applies .isxb timestamps to the Ethovision table while ignoring the first timestamp
to compensate for the length mismatch if the mismatch exists. More recent versions of Ethovision
should prevent this from occuring.</p>
</dd>
</dl>
<section id="id100">
<h2>Arguments<a class="headerlink" href="#id100" title="Link to this heading"></a></h2>
<dl>
<dt>input_ethovision_file: str</dt><dd><p>The path to the Ethovision file (.csv, .xlsx).</p>
</dd>
<dt>input_isxb_file: str</dt><dd><p>The path to the nVision (.isxb) movie file.</p>
</dd>
<dt>output_csv_file: str</dt><dd><p>The path to the output csv file.</p>
</dd>
<dt>input_ref_file: str | None</dt><dd><p>The path to the reference file (.isxd, .gpio, .isxb).
Timestamps from the .isxb file are aligned to the start time of this file.
Generally, this reference is a file from a miniscope recording which was synchronized with this .isxb file.
This argument is required if the <cite>time_ref</cite> is selected as ‘start’ or ‘unix’.
If the .isxb file is a standalone behaviour recording that is not synchronized to any miniscope file,
simply provide the .isxb file as the reference (if it’s required).</p>
</dd>
<dt>time_ref<span class="classifier">{‘start’, ‘unix’, ‘tsc’}</span></dt><dd><p>The time reference for the nVision (.isxb) timestamps.
If ‘tsc’ is used, the timestamps represent the hardware counter value in microseconds on the acquisition box when each frame was captured.
If ‘start’ is used, the timestamps represent the seconds since the start of the experiment.</p>
<blockquote>
<div><p>Note: <cite>input_ref_file</cite> must be specified if this option is used, otherwise an exception is thrown.
Timestamps are exported relative to the start time of the reference file.</p>
</div></blockquote>
<dl class="simple">
<dt>If ‘unix’ is used, the timestamps represent the seconds since the Unix epoch.</dt><dd><p>Note: <cite>input_ref_file</cite> must be specified if this option is used, otherwise an exception is thrown.
The <cite>input_ref_file</cite> ensures that <cite>isx.align_start_times</cite> has been called
on the .isxb file with the corresponding, synchronized miniscope file.</p>
</dd>
</dl>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_event_set_to_csv">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_event_set_to_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_event_set_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_csv_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_props_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sparse_output</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_amplitude</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_event_set_to_csv" title="Link to this definition"></a></dt>
<dd><p>Export event sets to a CSV file.</p>
<p>For more details see <span class="xref std std-ref">exportCellsAndStuff</span>.</p>
<section id="id101">
<h2>Arguments<a class="headerlink" href="#id101" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_event_set_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the cell sets to export.</p>
</dd>
<dt>output_csv_file<span class="classifier">str</span></dt><dd><p>The path of the CSV file to write.</p>
</dd>
<dt>time_ref<span class="classifier">{‘start’, ‘unix’}</span></dt><dd><p>The time reference for the CSV time stamps.
If ‘start’ is used, the time stamps represent the seconds since the start of the cell set.
If ‘unix’ is used, the time stamps represents the second since the Unix epoch.</p>
</dd>
<dt>output_props_file<span class="classifier">str</span></dt><dd><p>The path of the properties CSV file to write.</p>
</dd>
<dt>sparse_output: bool</dt><dd><p>If true, output events in sparse format showing all time points,
otherwise, output events in dense format showing only timepoints with events.</p>
</dd>
<dt>write_amplitude: bool</dt><dd><p>Only relevant when sparse_output is True.
If true, write amplitudes of each event,
otherwise, writes 1 where events occur and 0 elsewhere.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_gpio_set_to_csv">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_gpio_set_to_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_gpio_set_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_csv_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inter_isxd_file_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'/tmp'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_gpio_set_to_csv" title="Link to this definition"></a></dt>
<dd><p>Export gpio sets to a CSV file.</p>
<p>If exporting more than one file, for correct formatting, files should either be all non .imu files or all .imu files.</p>
<p>For more details see <span class="xref std std-ref">exportCellsAndStuff</span>.</p>
<section id="id102">
<h2>Arguments<a class="headerlink" href="#id102" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_gpio_set_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the cell sets to export.</p>
</dd>
<dt>output_csv_file<span class="classifier">str</span></dt><dd><p>The path of the CSV file to write.</p>
</dd>
<dt>inter_isxd_file_dir<span class="classifier">str</span></dt><dd><p>The path of the directory to put intermediate .isxd file
The default path for Mac &amp; Linux is /tmp. The default path for Windows is the directory containing the input gpio file.</p>
</dd>
<dt>time_ref<span class="classifier">{‘start’, ‘unix’}</span></dt><dd><p>The time reference for the CSV time stamps.
If ‘start’ is used, the time stamps represent the seconds since the start of the cell set.
If ‘unix’ is used, the time stamps represents the second since the Unix epoch.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_gpio_to_isxd">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_gpio_to_isxd</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_gpio_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_isxd_dir</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_gpio_to_isxd" title="Link to this definition"></a></dt>
<dd><p>Export GPIO file (.gpio, .raw, .hdf5, .imu) to an Inscopix Data File (.isxd).</p>
<p>Output file will have the name &lt;input_file_name&gt;_gpio.isxd</p>
<section id="id103">
<h2>Arguments<a class="headerlink" href="#id103" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_gpio_file<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file path of the gpio file to be exported.</p>
</dd>
<dt>output_isxd_dir<span class="classifier">str</span></dt><dd><p>The path of the directory to write isxd file.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_image_to_isxd_tiff">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_image_to_isxd_tiff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">isxd_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tiff_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">rgb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_image_to_isxd_tiff" title="Link to this definition"></a></dt>
<dd><p>Save an image as isxd and tiff.
If rgb is one of ‘red’, ‘green’, ‘blue’, the images will be colored.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_image_to_tiff">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_image_to_tiff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">image</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">tiff_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_rgb</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_image_to_tiff" title="Link to this definition"></a></dt>
<dd><p>Save an image as a uint16 tiff.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_isxd_image_to_tiff">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_isxd_image_to_tiff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_isxd_image</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_tiff_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_isxd_image_to_tiff" title="Link to this definition"></a></dt>
<dd><p>Convert an ISXD image file to a tiff image. 
WARNING: does not export tiff with alignment metadata, use align_image instead</p>
<section id="id104">
<h2>Arguments<a class="headerlink" href="#id104" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_isxd_image<span class="classifier">str</span></dt><dd><p>Path to the input ISXD image file.</p>
</dd>
<dt>output_tiff_file<span class="classifier">str</span></dt><dd><p>Path to the output tiff file.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_movie_timestamps_to_csv">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_movie_timestamps_to_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_movie_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_csv_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_movie_timestamps_to_csv" title="Link to this definition"></a></dt>
<dd><p>Export movie frame timestamps to a csv file.
This operation is supported for .isxd and .isxb movies.</p>
<p>For more details see <span class="xref std std-ref">exportMovie</span>.</p>
<section id="id105">
<h2>Arguments<a class="headerlink" href="#id105" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_movie_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the movies to export frame timestamps from.</p>
</dd>
<dt>output_csv_file<span class="classifier">str</span></dt><dd><p>The path of the csv file to be written.</p>
</dd>
<dt>time_ref<span class="classifier">{‘start’, ‘unix’, ‘tsc’}</span></dt><dd><p>The time reference for the CSV time stamps.
If ‘start’ is used, the timestamps represent the seconds since the start of the movie.
If ‘unix’ is used, the timestamps represent the seconds since the Unix epoch.
If ‘tsc’ is used, the timestamps represent the hardware counter value on the acquisition box when each frame was captured.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_movie_to_mp4">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_movie_to_mp4</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_movie_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_mp4_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compression_quality</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_invalid_frames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">frame_rate_format</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'float'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">draw_bounding_box</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">draw_bounding_box_center</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">draw_zones</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_movie_to_mp4" title="Link to this definition"></a></dt>
<dd><p>Export movies to an MP4 file.</p>
<p>For more details see <span class="xref std std-ref">exportMovie</span>.</p>
<section id="id106">
<h2>Arguments<a class="headerlink" href="#id106" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_movie_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the movies to be exported.</p>
</dd>
<dt>output_mp4_file<span class="classifier">str</span></dt><dd><p>The path of the MP4 file to be written.</p>
</dd>
<dt>compression_quality<span class="classifier">float</span></dt><dd><p>A value between 0.001 and 1 that controls the quality and size of the output file for the MP4 format.
The larger the value, the better the quality of the movie, but the larger the size of the file.
The default value of 0.1 typically produces a good quality movie where the file is at most 10% of the uncompressed original file size,
but may be smaller if the movie content can be efficiently encoded.
More formally, this represents a rough maximum on the output file size as a fraction of the original file size.</p>
</dd>
<dt>write_invalid_frames<span class="classifier">bool</span></dt><dd><p>If True, write invalid (dropped, cropped, and blank) frames as zero,
otherwise, do not write them at all.</p>
</dd>
<dt>frame_rate_format<span class="classifier">{“float”, “int”}</span></dt><dd><p>Format to encode the frame rate in the output mp4 file.
If float, the frame rate will be exported as a precise estimate of the input movie sampling rate.
If int, the frame rate will be rounded to the nearest integer.</p>
</dd>
<dt>draw_bounding_box<span class="classifier">bool</span></dt><dd><p>Only used for nVision <cite>.isxb</cite> movies.
If there is nVision tracking data in the <cite>.isxb</cite> movie, and this flag is enabled,
then draw the bounding box estimate on each frame of the exported mp4 movie.</p>
</dd>
<dt>draw_bounding_box_center<span class="classifier">bool</span></dt><dd><p>Only used for nVision <cite>.isxb</cite> movies.
If there is nVision tracking data in the <cite>.isxb</cite> movie, and this flag is enabled,
then draw the center of the bounding box estimate on each frame of the exported mp4 movie.</p>
</dd>
<dt>draw_zones<span class="classifier">bool</span></dt><dd><p>Only used for nVision <cite>.isxb</cite> movies.
If there is nVision tracking data in the <cite>.isxb</cite> movie, and this flag is enabled,
then draw the zones of the tracking area on each frame of the exported mp4 movie.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_movie_to_nwb">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_movie_to_nwb</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_movie_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_nwb_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">identifier</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">session_description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">comments</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">experiment_description</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">experimenter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">institution</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lab</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">session_id</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_movie_to_nwb" title="Link to this definition"></a></dt>
<dd><p>Export movies to an HDF5-based neurodata without borders (NWB) file.</p>
<p>For more details see <span class="xref std std-ref">exportMovie</span>.</p>
<section id="id107">
<h2>Arguments<a class="headerlink" href="#id107" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_movie_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the movies to be exported.</p>
</dd>
<dt>output_nwb_file<span class="classifier">str</span></dt><dd><p>The path of the NWB file to be written.</p>
</dd>
<dt>identifier<span class="classifier">str</span></dt><dd><p>An identifier for the file according to the NWB spec.</p>
</dd>
<dt>session_description<span class="classifier">str</span></dt><dd><p>A session description for the file according to the NWB spec.</p>
</dd>
<dt>comments<span class="classifier">str</span></dt><dd><p>Comments on the recording session.</p>
</dd>
<dt>description<span class="classifier">str</span></dt><dd><p>Description for the file according to the NWB spec.</p>
</dd>
<dt>experiment_description<span class="classifier">str</span></dt><dd><p>Details about the experiment.</p>
</dd>
<dt>experimenter<span class="classifier">str</span></dt><dd><p>The person who recorded the data.</p>
</dd>
<dt>institution<span class="classifier">str</span></dt><dd><p>The place where the recording was performed.</p>
</dd>
<dt>lab<span class="classifier">str</span></dt><dd><p>The lab where the recording was performed.</p>
</dd>
<dt>session_id<span class="classifier">str</span></dt><dd><p>A unique session identifier for the recording.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_movie_to_tiff">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_movie_to_tiff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_movie_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_tiff_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">write_invalid_frames</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_movie_to_tiff" title="Link to this definition"></a></dt>
<dd><p>Export movies to a TIFF file.</p>
<p>For more details see <span class="xref std std-ref">exportMovie</span>.</p>
<section id="id108">
<h2>Arguments<a class="headerlink" href="#id108" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_movie_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the movies to be exported.</p>
</dd>
<dt>output_tiff_file<span class="classifier">str</span></dt><dd><p>The path of the TIFF file to be written.</p>
</dd>
<dt>write_invalid_frames<span class="classifier">bool</span></dt><dd><p>If True, write invalid (dropped, cropped, and blank) frames as zero,
otherwise, do not write them at all.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_nvision_movie_tracking_frame_data_to_csv">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_nvision_movie_tracking_frame_data_to_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_movie_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_csv_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_nvision_movie_tracking_frame_data_to_csv" title="Link to this definition"></a></dt>
<dd><p>Export frame tracking metadata from an nVision movie.
This operation is supported for .isxb movies.</p>
<dl class="simple">
<dt>The frame tracking metadata is generated by the nVision tracking model, and includes the following columns:</dt><dd><ul class="simple">
<li><p>Global Frame Number: The frame number in the input movie series.</p></li>
<li><p>Movie Number: The movie number in the series.</p></li>
<li><p>Local Frame Number: The frame number in the individual movie.</p></li>
<li><p>Frame Timestamp: The frame timestamp. In units of seconds or microseconds based on the input <cite>time_ref</cite> parameter.</p></li>
<li><p>Bounding Box Left: X coordinate of the top left corner of the bounding box. In units of pixels.</p></li>
<li><p>Bounding Box Top: Y coordinate of the top left corner of the bounding box. In units of pixels.</p></li>
<li><p>Bounding Box Right: X coordinate of the bottom right corner of the bounding box. In units of pixels.</p></li>
<li><p>Bounding Box Bottom: Y coordinate of the bottom right corner of the bounding box. In units of pixels.</p></li>
<li><p>Bounding Box Center X: X coordinate of the center point of the bounding box. In units of pixels.</p></li>
<li><p>Bounding Box Center Y: Y coordinate of the center point of the bounding box. In units of pixels.</p></li>
<li><p>Confidence: The nVision tracking model confidence of the bounding box estimate. In units of %.</p></li>
<li><p>Zone ID: If the nVision tracking model detected the mouse was inside a zone, this column contains the id of that zone.</p></li>
<li><p>Zone Name: If the nVision tracking model detected the mouse was inside a zone, this column contains the name of that zone.</p></li>
</ul>
</dd>
</dl>
<p>For more details see <span class="xref std std-ref">exportMovie</span>.</p>
<section id="id109">
<h2>Arguments<a class="headerlink" href="#id109" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_movie_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the movies to export frame tracking metadata from.</p>
</dd>
<dt>output_csv_file<span class="classifier">str</span></dt><dd><p>The path of the csv file to be written.</p>
</dd>
<dt>time_ref<span class="classifier">{‘start’, ‘unix’, ‘tsc’}</span></dt><dd><p>The time reference for the CSV time stamps.
If ‘start’ is used, the timestamps represent the seconds since the start of the movie.
If ‘unix’ is used, the timestamps represent the seconds since the Unix epoch.
If ‘tsc’ is used, the timestamps represent the hardware counter value on the acquisition box when each frame was captured.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_nvision_movie_tracking_zone_data_to_csv">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_nvision_movie_tracking_zone_data_to_csv</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_movie_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_csv_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_nvision_movie_tracking_zone_data_to_csv" title="Link to this definition"></a></dt>
<dd><p>Export zone tracking metadata from an nVision movie.
This operation is supported for .isxb movies.</p>
<dl class="simple">
<dt>The zone tracking metadata used by the nVision tracking model, and includes the following columns:</dt><dd><ul class="simple">
<li><p>ID: Unique identifier for the zone.</p></li>
<li><p>Enabled: Flag indicating whether the zone is enabled for tracking.</p></li>
<li><p>Name: User friendly name for the zone.</p></li>
<li><p>Description: Optional description for the zone</p></li>
<li><p>Type: The shape of the zone. Can be either rectangle, polygon, or ellipse.</p></li>
<li><dl class="simple">
<dt>X <cite>i</cite>: The ith X coordinate of the zone.</dt><dd><p>Note: Since zones can have different shapes, they can have a different number of coordinates.
The csv output will contain <cite>n</cite> columns for the X coordinate, where <cite>n</cite> is the maxmimum
number of coordinates across all zones in the metadata.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Y <cite>i</cite>: The ith Y coordinate of the zone.</dt><dd><p>Note: Since zones can have different shapes, they can have a different number of coordinates.
The csv output will contain <cite>n</cite> columns for the Y coordinate, where <cite>n</cite> is the maxmimum
number of coordinates across all zones in the metadata.</p>
</dd>
</dl>
</li>
<li><p>Major Axis: only outputted for ellipse shaped zones. Length of the major axis.</p></li>
<li><p>Minor Axis: only outputted for ellipse shaped zones. Length of the minor axis.</p></li>
<li><p>Angle: only outputted for ellipse shaped zones. Ellipse rotation angle in degrees.</p></li>
</ul>
</dd>
</dl>
<p>For more details see <span class="xref std std-ref">exportMovie</span>.</p>
<section id="id110">
<h2>Arguments<a class="headerlink" href="#id110" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_movie_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the movies to export frame tracking metadata from.</p>
</dd>
<dt>output_csv_file<span class="classifier">str</span></dt><dd><p>The path of the csv file to be written.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_tiff_to_isxd_image">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_tiff_to_isxd_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_tiff_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_isxd_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_tiff_to_isxd_image" title="Link to this definition"></a></dt>
<dd><p>Convert an tiff image file to an ISXD image. 
WARNING: does not export tiff with alignment metadata, use align_image instead</p>
<section id="id111">
<h2>Arguments<a class="headerlink" href="#id111" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_tiff_file<span class="classifier">str</span></dt><dd><p>Path to the input tiff file.</p>
</dd>
<dt>output_isxd_file<span class="classifier">str</span></dt><dd><p>Path to the output ISXD image file.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.export_vessel_set_to_csv_tiff">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">export_vessel_set_to_csv_tiff</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_vessel_set_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_trace_csv_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_line_csv_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_map_tiff_file</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_heatmaps_tiff_dir</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">time_ref</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'start'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.export_vessel_set_to_csv_tiff" title="Link to this definition"></a></dt>
<dd><p>Export vessel sets to a CSV file with trace data and TIFF files with image data.</p>
<section id="id112">
<h2>Arguments<a class="headerlink" href="#id112" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_vessel_set_files<span class="classifier">list&lt;str&gt;</span></dt><dd><p>The file paths of the vessel sets to export.</p>
</dd>
<dt>output_trace_csv_file<span class="classifier">str</span></dt><dd><p>The path of the trace CSV file to write.</p>
</dd>
<dt>output_line_csv_file<span class="classifier">str</span></dt><dd><p>The path of the line CSV file to write.</p>
</dd>
<dt>output_map_tiff_file<span class="classifier">str</span></dt><dd><p>The name of the vessel map TIFF file to write.</p>
</dd>
<dt>output_heatmaps_tiff_dir<span class="classifier">str</span></dt><dd><p>The name of the directory to write correlation heatmaps as TIFF stacks.</p>
</dd>
<dt>time_ref<span class="classifier">{‘start’, ‘unix’}</span></dt><dd><p>The time reference for the CSV time stamps.
If ‘start’ is used, the time stamps represent the seconds since the start of the vessel set.
If ‘unix’ is used, the time stamps represents the second since the Unix epoch.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.get_file_extension">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">get_file_extension</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.get_file_extension" title="Link to this definition"></a></dt>
<dd><p>Get full extension for file that can have multiple extensions.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.get_file_stem">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">get_file_stem</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.get_file_stem" title="Link to this definition"></a></dt>
<dd><p>Get the file stem for file that can have multiple extensions.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.make_output_file_path">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">make_output_file_path</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ext</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'isxd'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.make_output_file_path" title="Link to this definition"></a></dt>
<dd><p>Make an output file path from an input path, output directory, suffix and extension.</p>
<p>This is useful for generate output file paths for processing steps.</p>
<section id="id113">
<h2>Arguments<a class="headerlink" href="#id113" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>in_file<span class="classifier">str</span></dt><dd><p>The input file path.</p>
</dd>
<dt>out_dir<span class="classifier">str</span></dt><dd><p>The output directory path.</p>
</dd>
<dt>suffix<span class="classifier">str</span></dt><dd><p>The suffix to append to the file stem with a ‘-‘.
If left empty, no suffix is appended.</p>
</dd>
<dt>ext<span class="classifier">‘isxd’</span></dt><dd><p>The output file extension, not including the ‘.’.</p>
</dd>
</dl>
</section>
<section id="id114">
<h2>Returns<a class="headerlink" href="#id114" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>str</dt><dd><p>The output file path.</p>
</dd>
</dl>
</section>
<section id="id115">
<h2>Examples<a class="headerlink" href="#id115" title="Link to this heading"></a></h2>
<p>Make the output file path for a preprocessed recording.</p>
<div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">make_output_file_path</span><span class="p">(</span><span class="s1">&#39;in_dir/in_file.xml&#39;</span><span class="p">,</span> <span class="s1">&#39;out_dir&#39;</span><span class="p">,</span> <span class="s1">&#39;PP&#39;</span><span class="p">)</span>
<span class="go">&#39;out_dir/in_file-PP.isxd&#39;</span>
</pre></div>
</div>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.make_output_file_paths">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">make_output_file_paths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_files</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">out_dir</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">suffix</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ext</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'isxd'</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.make_output_file_paths" title="Link to this definition"></a></dt>
<dd><p>Like <a class="reference internal" href="#isx.make_output_file_path" title="isx.make_output_file_path"><code class="xref py py-func docutils literal notranslate"><span class="pre">isx.make_output_file_path()</span></code></a>, but for many files.</p>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.overlay_cell_map_on_image">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">overlay_cell_map_on_image</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_isxd_cell_map_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">input_isxd_image_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">output_tiff_image_file</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.overlay_cell_map_on_image" title="Link to this definition"></a></dt>
<dd><p>Overlay a cellmap onto an image, and save as isxd. Can save tiff as well.</p>
<section id="id116">
<h2>Arguments<a class="headerlink" href="#id116" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>input_isxd_cell_map_file<span class="classifier">str</span></dt><dd><p>Path to an .isxd cell map image file.</p>
</dd>
<dt>input_isxd_image_file<span class="classifier">str</span></dt><dd><p>Path to an .isxd image file to overlay on.</p>
</dd>
<dt>output_tiff_image_file<span class="classifier">str</span></dt><dd><p>Path to the output isxd cell map image.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.overlay_cellmaps">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">overlay_cellmaps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">first_tiff_cellmap_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_tiff_cellmap_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overlayed_tiff_cellmap_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">background_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#000000'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">first_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#00ff00'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">second_color</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'#ff00ff'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cell_thresh</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0.5</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.overlay_cellmaps" title="Link to this definition"></a></dt>
<dd><p>Overlay two cellmaps using different colors to show overlap.</p>
<section id="id117">
<h2>Arguments<a class="headerlink" href="#id117" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>first_tiff_cellmap_file<span class="classifier">str</span></dt><dd><p>Path to the first tiff cellmap image.</p>
</dd>
<dt>second_tiff_cellmap_file<span class="classifier">str</span></dt><dd><p>Path to the second tiff cellmap image.</p>
</dd>
<dt>overlayed_tiff_cellmap_file<span class="classifier">str</span></dt><dd><p>Path to the output tiff cellmap.</p>
</dd>
<dt>background_color<span class="classifier">str</span></dt><dd><p>Hex color code for background. Format: #RRGGBB</p>
</dd>
<dt>first_color<span class="classifier">str</span></dt><dd><p>Hex color code for cells in first cellmap. Format: #RRGGBB</p>
</dd>
<dt>second_color<span class="classifier">str</span></dt><dd><p>Hex color code for cells in second cellmap. Format: #RRGGBB</p>
</dd>
<dt>cell_thresh<span class="classifier">float [0, 1]</span></dt><dd><p>Pixel values less than cell_thresh will be considered as the background.</p>
</dd>
</dl>
</section>
</dd></dl>

<dl class="py function">
<dt class="sig sig-object py" id="isx.verify_deinterleave">
<span class="sig-prename descclassname"><span class="pre">isx.</span></span><span class="sig-name descname"><span class="pre">verify_deinterleave</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">in_file</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">efocus</span></span></em><span class="sig-paren">)</span><a class="headerlink" href="#isx.verify_deinterleave" title="Link to this definition"></a></dt>
<dd><p>Verify if all frames from movie has the same efocus as provided.</p>
<section id="id118">
<h2>Arguments<a class="headerlink" href="#id118" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>in_file<span class="classifier">str</span></dt><dd><p>The input file path.</p>
</dd>
<dt>efocus<span class="classifier">int</span></dt><dd><p>The efocus value to be compared with.</p>
</dd>
</dl>
</section>
<section id="id119">
<h2>Returns<a class="headerlink" href="#id119" title="Link to this heading"></a></h2>
<dl class="simple">
<dt>bool</dt><dd><p>True if the movie is successfully verified, False otherwise.</p>
</dd>
</dl>
</section>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="examples.html" class="btn btn-neutral float-right" title="Examples" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright Inscopix, Inc..</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>